{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from copy import deepcopy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from scipy.interpolate import CubicSpline\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import warnings\n",
    "import itertools\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "import statsmodels.stats.api as sms\n",
    "import scipy.stats.kde as kde"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read 16S data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Abiotrophia</th>\n",
       "      <th>Acetanaerobacterium</th>\n",
       "      <th>Acidaminococcus</th>\n",
       "      <th>Actinidia_dash_chinensis</th>\n",
       "      <th>Actinomyces</th>\n",
       "      <th>Actinomyces_dash_massiliensis</th>\n",
       "      <th>Adlercreutzia</th>\n",
       "      <th>Agathobacter</th>\n",
       "      <th>Akkermansia</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "      <th>bacterium_dash_YE57</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_BBMC_dash_9</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_Ecwsrb026</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_RB7c</th>\n",
       "      <th>butyrate_dash_producing_dash_bacterium_dash_L2_dash_10</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>vadinBE97</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6185224</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007017</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>...</td>\n",
       "      <td>0.030629</td>\n",
       "      <td>0.000628</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000221</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.038717</td>\n",
       "      <td>0.001327</td>\n",
       "      <td>0.044047</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185259</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008313</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001122</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023107</td>\n",
       "      <td>0.000313</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.021788</td>\n",
       "      <td>0.001451</td>\n",
       "      <td>0.069750</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185212</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000855</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001043</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015071</td>\n",
       "      <td>0.000540</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.031864</td>\n",
       "      <td>0.002363</td>\n",
       "      <td>0.049940</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185177</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000279</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001221</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001117</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023953</td>\n",
       "      <td>0.001376</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000290</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.042888</td>\n",
       "      <td>0.001821</td>\n",
       "      <td>0.050959</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185278</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.083783</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001688</td>\n",
       "      <td>0.000763</td>\n",
       "      <td>0.039542</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 469 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Abiotrophia  Acetanaerobacterium  Acidaminococcus  \\\n",
       "SampleID                                                        \n",
       "SRR6185224          0.0                  0.0              0.0   \n",
       "SRR6185259          0.0                  0.0              0.0   \n",
       "SRR6185212          0.0                  0.0              0.0   \n",
       "SRR6185177          0.0                  0.0              0.0   \n",
       "SRR6185278          0.0                  0.0              0.0   \n",
       "\n",
       "            Actinidia_dash_chinensis  Actinomyces  \\\n",
       "SampleID                                            \n",
       "SRR6185224                  0.000000          0.0   \n",
       "SRR6185259                  0.000000          0.0   \n",
       "SRR6185212                  0.000000          0.0   \n",
       "SRR6185177                  0.000279          0.0   \n",
       "SRR6185278                  0.000000          0.0   \n",
       "\n",
       "            Actinomyces_dash_massiliensis  Adlercreutzia  Agathobacter  \\\n",
       "SampleID                                                                 \n",
       "SRR6185224                            0.0            0.0      0.007017   \n",
       "SRR6185259                            0.0            0.0      0.008313   \n",
       "SRR6185212                            0.0            0.0      0.000855   \n",
       "SRR6185177                            0.0            0.0      0.001221   \n",
       "SRR6185278                            0.0            0.0      0.083783   \n",
       "\n",
       "            Akkermansia  Akkermansia_dash_muciniphila  ...  \\\n",
       "SampleID                                               ...   \n",
       "SRR6185224          0.0                      0.000337  ...   \n",
       "SRR6185259          0.0                      0.001122  ...   \n",
       "SRR6185212          0.0                      0.001043  ...   \n",
       "SRR6185177          0.0                      0.001117  ...   \n",
       "SRR6185278          0.0                      0.000000  ...   \n",
       "\n",
       "            _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \\\n",
       "SampleID                                                                                  \n",
       "SRR6185224                                           0.030629                             \n",
       "SRR6185259                                           0.023107                             \n",
       "SRR6185212                                           0.015071                             \n",
       "SRR6185177                                           0.023953                             \n",
       "SRR6185278                                           0.000203                             \n",
       "\n",
       "            bacterium_dash_YE57  \\\n",
       "SampleID                          \n",
       "SRR6185224             0.000628   \n",
       "SRR6185259             0.000313   \n",
       "SRR6185212             0.000540   \n",
       "SRR6185177             0.001376   \n",
       "SRR6185278             0.000000   \n",
       "\n",
       "            bacterium_dash_enrichment_dash_culture_dash_clone_dash_BBMC_dash_9  \\\n",
       "SampleID                                                                         \n",
       "SRR6185224                                                0.0                    \n",
       "SRR6185259                                                0.0                    \n",
       "SRR6185212                                                0.0                    \n",
       "SRR6185177                                                0.0                    \n",
       "SRR6185278                                                0.0                    \n",
       "\n",
       "            bacterium_dash_enrichment_dash_culture_dash_clone_dash_Ecwsrb026  \\\n",
       "SampleID                                                                       \n",
       "SRR6185224                                           0.000221                  \n",
       "SRR6185259                                           0.000181                  \n",
       "SRR6185212                                           0.000000                  \n",
       "SRR6185177                                           0.000290                  \n",
       "SRR6185278                                           0.000000                  \n",
       "\n",
       "            bacterium_dash_enrichment_dash_culture_dash_clone_dash_RB7c  \\\n",
       "SampleID                                                                  \n",
       "SRR6185224                                                0.0             \n",
       "SRR6185259                                                0.0             \n",
       "SRR6185212                                                0.0             \n",
       "SRR6185177                                                0.0             \n",
       "SRR6185278                                                0.0             \n",
       "\n",
       "            butyrate_dash_producing_dash_bacterium_dash_L2_dash_10  \\\n",
       "SampleID                                                             \n",
       "SRR6185224                                                0.0        \n",
       "SRR6185259                                                0.0        \n",
       "SRR6185212                                                0.0        \n",
       "SRR6185177                                                0.0        \n",
       "SRR6185278                                                0.0        \n",
       "\n",
       "            gut_dash_metagenome  human_dash_gut_dash_metagenome  metagenome  \\\n",
       "SampleID                                                                      \n",
       "SRR6185224             0.038717                        0.001327    0.044047   \n",
       "SRR6185259             0.021788                        0.001451    0.069750   \n",
       "SRR6185212             0.031864                        0.002363    0.049940   \n",
       "SRR6185177             0.042888                        0.001821    0.050959   \n",
       "SRR6185278             0.001688                        0.000763    0.039542   \n",
       "\n",
       "            vadinBE97  \n",
       "SampleID               \n",
       "SRR6185224        0.0  \n",
       "SRR6185259        0.0  \n",
       "SRR6185212        0.0  \n",
       "SRR6185177        0.0  \n",
       "SRR6185278        0.0  \n",
       "\n",
       "[5 rows x 469 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_16S = pd.read_excel('./16S_relative_abundance.xlsx', index_col=0)\n",
    "df_16S.columns = [c.replace('/','_slash_').replace(' ','_space_').replace('[','_leftsquarebracket_').replace(']','_rightsquarebracket_').replace('-','_dash_').replace('.','_dot_').replace('(','_leftroundbracket').replace(')','_rightroundbracket_') for c in df_16S.columns]\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add pseudocount for zeros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Abiotrophia</th>\n",
       "      <th>Acetanaerobacterium</th>\n",
       "      <th>Acidaminococcus</th>\n",
       "      <th>Actinidia_dash_chinensis</th>\n",
       "      <th>Actinomyces</th>\n",
       "      <th>Actinomyces_dash_massiliensis</th>\n",
       "      <th>Adlercreutzia</th>\n",
       "      <th>Agathobacter</th>\n",
       "      <th>Akkermansia</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "      <th>bacterium_dash_YE57</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_BBMC_dash_9</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_Ecwsrb026</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_RB7c</th>\n",
       "      <th>butyrate_dash_producing_dash_bacterium_dash_L2_dash_10</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>vadinBE97</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6185224</th>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.007017</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>...</td>\n",
       "      <td>0.030629</td>\n",
       "      <td>0.000628</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000221</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.038717</td>\n",
       "      <td>0.001327</td>\n",
       "      <td>0.044047</td>\n",
       "      <td>0.000047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185259</th>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.008313</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.001122</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023107</td>\n",
       "      <td>0.000313</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.021788</td>\n",
       "      <td>0.001451</td>\n",
       "      <td>0.069750</td>\n",
       "      <td>0.000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185212</th>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000855</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.001043</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015071</td>\n",
       "      <td>0.000540</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.031864</td>\n",
       "      <td>0.002363</td>\n",
       "      <td>0.049940</td>\n",
       "      <td>0.000025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185177</th>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000279</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.001221</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.001117</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023953</td>\n",
       "      <td>0.001376</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000290</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.042888</td>\n",
       "      <td>0.001821</td>\n",
       "      <td>0.050959</td>\n",
       "      <td>0.000031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185278</th>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.083783</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.001688</td>\n",
       "      <td>0.000763</td>\n",
       "      <td>0.039542</td>\n",
       "      <td>0.000041</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 469 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Abiotrophia  Acetanaerobacterium  Acidaminococcus  \\\n",
       "SampleID                                                        \n",
       "SRR6185224     0.000047             0.000047         0.000047   \n",
       "SRR6185259     0.000033             0.000033         0.000033   \n",
       "SRR6185212     0.000025             0.000025         0.000025   \n",
       "SRR6185177     0.000031             0.000031         0.000031   \n",
       "SRR6185278     0.000041             0.000041         0.000041   \n",
       "\n",
       "            Actinidia_dash_chinensis  Actinomyces  \\\n",
       "SampleID                                            \n",
       "SRR6185224                  0.000047     0.000047   \n",
       "SRR6185259                  0.000033     0.000033   \n",
       "SRR6185212                  0.000025     0.000025   \n",
       "SRR6185177                  0.000279     0.000031   \n",
       "SRR6185278                  0.000041     0.000041   \n",
       "\n",
       "            Actinomyces_dash_massiliensis  Adlercreutzia  Agathobacter  \\\n",
       "SampleID                                                                 \n",
       "SRR6185224                       0.000047       0.000047      0.007017   \n",
       "SRR6185259                       0.000033       0.000033      0.008313   \n",
       "SRR6185212                       0.000025       0.000025      0.000855   \n",
       "SRR6185177                       0.000031       0.000031      0.001221   \n",
       "SRR6185278                       0.000041       0.000041      0.083783   \n",
       "\n",
       "            Akkermansia  Akkermansia_dash_muciniphila  ...  \\\n",
       "SampleID                                               ...   \n",
       "SRR6185224     0.000047                      0.000337  ...   \n",
       "SRR6185259     0.000033                      0.001122  ...   \n",
       "SRR6185212     0.000025                      0.001043  ...   \n",
       "SRR6185177     0.000031                      0.001117  ...   \n",
       "SRR6185278     0.000041                      0.000041  ...   \n",
       "\n",
       "            _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \\\n",
       "SampleID                                                                                  \n",
       "SRR6185224                                           0.030629                             \n",
       "SRR6185259                                           0.023107                             \n",
       "SRR6185212                                           0.015071                             \n",
       "SRR6185177                                           0.023953                             \n",
       "SRR6185278                                           0.000203                             \n",
       "\n",
       "            bacterium_dash_YE57  \\\n",
       "SampleID                          \n",
       "SRR6185224             0.000628   \n",
       "SRR6185259             0.000313   \n",
       "SRR6185212             0.000540   \n",
       "SRR6185177             0.001376   \n",
       "SRR6185278             0.000041   \n",
       "\n",
       "            bacterium_dash_enrichment_dash_culture_dash_clone_dash_BBMC_dash_9  \\\n",
       "SampleID                                                                         \n",
       "SRR6185224                                           0.000047                    \n",
       "SRR6185259                                           0.000033                    \n",
       "SRR6185212                                           0.000025                    \n",
       "SRR6185177                                           0.000031                    \n",
       "SRR6185278                                           0.000041                    \n",
       "\n",
       "            bacterium_dash_enrichment_dash_culture_dash_clone_dash_Ecwsrb026  \\\n",
       "SampleID                                                                       \n",
       "SRR6185224                                           0.000221                  \n",
       "SRR6185259                                           0.000181                  \n",
       "SRR6185212                                           0.000025                  \n",
       "SRR6185177                                           0.000290                  \n",
       "SRR6185278                                           0.000041                  \n",
       "\n",
       "            bacterium_dash_enrichment_dash_culture_dash_clone_dash_RB7c  \\\n",
       "SampleID                                                                  \n",
       "SRR6185224                                           0.000047             \n",
       "SRR6185259                                           0.000033             \n",
       "SRR6185212                                           0.000025             \n",
       "SRR6185177                                           0.000031             \n",
       "SRR6185278                                           0.000041             \n",
       "\n",
       "            butyrate_dash_producing_dash_bacterium_dash_L2_dash_10  \\\n",
       "SampleID                                                             \n",
       "SRR6185224                                           0.000047        \n",
       "SRR6185259                                           0.000033        \n",
       "SRR6185212                                           0.000025        \n",
       "SRR6185177                                           0.000031        \n",
       "SRR6185278                                           0.000041        \n",
       "\n",
       "            gut_dash_metagenome  human_dash_gut_dash_metagenome  metagenome  \\\n",
       "SampleID                                                                      \n",
       "SRR6185224             0.038717                        0.001327    0.044047   \n",
       "SRR6185259             0.021788                        0.001451    0.069750   \n",
       "SRR6185212             0.031864                        0.002363    0.049940   \n",
       "SRR6185177             0.042888                        0.001821    0.050959   \n",
       "SRR6185278             0.001688                        0.000763    0.039542   \n",
       "\n",
       "            vadinBE97  \n",
       "SampleID               \n",
       "SRR6185224   0.000047  \n",
       "SRR6185259   0.000033  \n",
       "SRR6185212   0.000025  \n",
       "SRR6185177   0.000031  \n",
       "SRR6185278   0.000041  \n",
       "\n",
       "[5 rows x 469 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for sample_id in df_16S.index:\n",
    "    sample = np.array(df_16S.loc[sample_id])\n",
    "    minval = np.min(sample[np.nonzero(sample)]) # minimum non-zero value\n",
    "    sample[sample==0] = minval\n",
    "    df_16S.loc[sample_id] = sample\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read metadata for inulin and control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Collection_Date</th>\n",
       "      <th>env_material</th>\n",
       "      <th>ETHNICITY</th>\n",
       "      <th>Host_Age</th>\n",
       "      <th>host_body_mass_index</th>\n",
       "      <th>Host_diet</th>\n",
       "      <th>host_sex</th>\n",
       "      <th>Group</th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>status</th>\n",
       "      <th>Replicate</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6185224</th>\n",
       "      <td>16-Apr</td>\n",
       "      <td>Feces</td>\n",
       "      <td>NZ European</td>\n",
       "      <td>52</td>\n",
       "      <td>20.6</td>\n",
       "      <td>High_dietary_fibre_intake</td>\n",
       "      <td>female</td>\n",
       "      <td>Prebiotic</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>pre</td>\n",
       "      <td>ADAPT03 sample 1</td>\n",
       "      <td>ADAPT03_STP1</td>\n",
       "      <td>Pre-intervention</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185259</th>\n",
       "      <td>16-Apr</td>\n",
       "      <td>Feces</td>\n",
       "      <td>NZ European</td>\n",
       "      <td>52</td>\n",
       "      <td>20.6</td>\n",
       "      <td>High_dietary_fibre_intake</td>\n",
       "      <td>female</td>\n",
       "      <td>Prebiotic</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>post</td>\n",
       "      <td>ADAPT03 sample 2</td>\n",
       "      <td>ADAPT03_ETP1</td>\n",
       "      <td>Post-intervention</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185212</th>\n",
       "      <td>16-May</td>\n",
       "      <td>Feces</td>\n",
       "      <td>NZ European</td>\n",
       "      <td>52</td>\n",
       "      <td>20.6</td>\n",
       "      <td>High_dietary_fibre_intake</td>\n",
       "      <td>female</td>\n",
       "      <td>Placebo</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>pre</td>\n",
       "      <td>ADAPT03 sample 3</td>\n",
       "      <td>ADAPT03_STP2</td>\n",
       "      <td>Pre-intervention</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185177</th>\n",
       "      <td>16-Jun</td>\n",
       "      <td>Feces</td>\n",
       "      <td>NZ European</td>\n",
       "      <td>52</td>\n",
       "      <td>20.6</td>\n",
       "      <td>High_dietary_fibre_intake</td>\n",
       "      <td>female</td>\n",
       "      <td>Placebo</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>post</td>\n",
       "      <td>ADAPT03 sample 4</td>\n",
       "      <td>ADAPT03_ETP2</td>\n",
       "      <td>Post-intervention</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185278</th>\n",
       "      <td>16-Apr</td>\n",
       "      <td>Feces</td>\n",
       "      <td>Chinese</td>\n",
       "      <td>52</td>\n",
       "      <td>22.3</td>\n",
       "      <td>High_dietary_fibre_intake</td>\n",
       "      <td>female</td>\n",
       "      <td>Placebo</td>\n",
       "      <td>ADAPT04</td>\n",
       "      <td>pre</td>\n",
       "      <td>ADAPT04 sample 1</td>\n",
       "      <td>ADAPT04_STP1</td>\n",
       "      <td>Pre-intervention</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Collection_Date env_material    ETHNICITY  Host_Age  \\\n",
       "#SampleID                                                        \n",
       "SRR6185224          16-Apr        Feces  NZ European        52   \n",
       "SRR6185259          16-Apr        Feces  NZ European        52   \n",
       "SRR6185212          16-May        Feces  NZ European        52   \n",
       "SRR6185177          16-Jun        Feces  NZ European        52   \n",
       "SRR6185278          16-Apr        Feces      Chinese        52   \n",
       "\n",
       "            host_body_mass_index                  Host_diet host_sex  \\\n",
       "#SampleID                                                              \n",
       "SRR6185224                  20.6  High_dietary_fibre_intake   female   \n",
       "SRR6185259                  20.6  High_dietary_fibre_intake   female   \n",
       "SRR6185212                  20.6  High_dietary_fibre_intake   female   \n",
       "SRR6185177                  20.6  High_dietary_fibre_intake   female   \n",
       "SRR6185278                  22.3  High_dietary_fibre_intake   female   \n",
       "\n",
       "                Group Subject_ID status         Replicate   Sample_Name  \\\n",
       "#SampleID                                                                 \n",
       "SRR6185224  Prebiotic    ADAPT03    pre  ADAPT03 sample 1  ADAPT03_STP1   \n",
       "SRR6185259  Prebiotic    ADAPT03   post  ADAPT03 sample 2  ADAPT03_ETP1   \n",
       "SRR6185212    Placebo    ADAPT03    pre  ADAPT03 sample 3  ADAPT03_STP2   \n",
       "SRR6185177    Placebo    ADAPT03   post  ADAPT03 sample 4  ADAPT03_ETP2   \n",
       "SRR6185278    Placebo    ADAPT04    pre  ADAPT04 sample 1  ADAPT04_STP1   \n",
       "\n",
       "                  SAMPLE_TYPE  \n",
       "#SampleID                      \n",
       "SRR6185224   Pre-intervention  \n",
       "SRR6185259  Post-intervention  \n",
       "SRR6185212   Pre-intervention  \n",
       "SRR6185177  Post-intervention  \n",
       "SRR6185278   Pre-intervention  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv('./metadata.txt', sep='\\t', index_col=0)\n",
    "df_meta = df_meta[['Host_die']]\n",
    "#df_meta = df_meta[(df_meta.Diet=='Inulin') | (df_meta.Diet=='Control')]\n",
    "#df_meta['Diet'] = [1 if g=='Inulin' else 0 for g in df_meta['Diet']]\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the 20 most abundant taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Muribaculaceae</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibaculum</th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Parabacteroides_dash_goldsteinii</th>\n",
       "      <th>Parasutterella</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Alloprevotella</th>\n",
       "      <th>Lachnospiraceae_dash_NK4A136_dash_group</th>\n",
       "      <th>Lachnospiraceae_dash_bacterium_dash_28_dash_4</th>\n",
       "      <th>Desulfovibrionaceae</th>\n",
       "      <th>Rikenellaceae_dash_RC9_dash_gut_dash_group</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Clostridia_dash_vadinBB60_dash_group</th>\n",
       "      <th>Bacteroides_dash_uniformis</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Helicobacter_dash_ganmani</th>\n",
       "      <th>Mucispirillum_dash_schaedleri</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>0.237934</td>\n",
       "      <td>0.004178</td>\n",
       "      <td>0.237536</td>\n",
       "      <td>0.018370</td>\n",
       "      <td>0.001134</td>\n",
       "      <td>0.024167</td>\n",
       "      <td>0.041179</td>\n",
       "      <td>0.036914</td>\n",
       "      <td>0.106118</td>\n",
       "      <td>0.029702</td>\n",
       "      <td>0.001105</td>\n",
       "      <td>0.001987</td>\n",
       "      <td>0.074031</td>\n",
       "      <td>0.006824</td>\n",
       "      <td>0.005138</td>\n",
       "      <td>0.000756</td>\n",
       "      <td>0.001861</td>\n",
       "      <td>0.011623</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>0.021927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>0.217026</td>\n",
       "      <td>0.022603</td>\n",
       "      <td>0.148081</td>\n",
       "      <td>0.038074</td>\n",
       "      <td>0.000506</td>\n",
       "      <td>0.030693</td>\n",
       "      <td>0.022308</td>\n",
       "      <td>0.038992</td>\n",
       "      <td>0.002435</td>\n",
       "      <td>0.031797</td>\n",
       "      <td>0.067491</td>\n",
       "      <td>0.008470</td>\n",
       "      <td>0.029207</td>\n",
       "      <td>0.033602</td>\n",
       "      <td>0.001221</td>\n",
       "      <td>0.000856</td>\n",
       "      <td>0.000614</td>\n",
       "      <td>0.010656</td>\n",
       "      <td>0.000101</td>\n",
       "      <td>0.114954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>0.236170</td>\n",
       "      <td>0.013355</td>\n",
       "      <td>0.213967</td>\n",
       "      <td>0.022921</td>\n",
       "      <td>0.000792</td>\n",
       "      <td>0.081351</td>\n",
       "      <td>0.063828</td>\n",
       "      <td>0.012851</td>\n",
       "      <td>0.017804</td>\n",
       "      <td>0.012587</td>\n",
       "      <td>0.003566</td>\n",
       "      <td>0.002575</td>\n",
       "      <td>0.086568</td>\n",
       "      <td>0.009055</td>\n",
       "      <td>0.002996</td>\n",
       "      <td>0.000850</td>\n",
       "      <td>0.000949</td>\n",
       "      <td>0.018225</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>0.069399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>0.181429</td>\n",
       "      <td>0.019500</td>\n",
       "      <td>0.116534</td>\n",
       "      <td>0.048061</td>\n",
       "      <td>0.000864</td>\n",
       "      <td>0.023371</td>\n",
       "      <td>0.106147</td>\n",
       "      <td>0.038334</td>\n",
       "      <td>0.017331</td>\n",
       "      <td>0.019330</td>\n",
       "      <td>0.000735</td>\n",
       "      <td>0.010890</td>\n",
       "      <td>0.108460</td>\n",
       "      <td>0.023867</td>\n",
       "      <td>0.000911</td>\n",
       "      <td>0.000592</td>\n",
       "      <td>0.000326</td>\n",
       "      <td>0.037851</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.023860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>0.225473</td>\n",
       "      <td>0.011760</td>\n",
       "      <td>0.218809</td>\n",
       "      <td>0.053997</td>\n",
       "      <td>0.000817</td>\n",
       "      <td>0.035780</td>\n",
       "      <td>0.053453</td>\n",
       "      <td>0.014144</td>\n",
       "      <td>0.020514</td>\n",
       "      <td>0.012380</td>\n",
       "      <td>0.000686</td>\n",
       "      <td>0.003082</td>\n",
       "      <td>0.120364</td>\n",
       "      <td>0.006152</td>\n",
       "      <td>0.000512</td>\n",
       "      <td>0.000283</td>\n",
       "      <td>0.006032</td>\n",
       "      <td>0.032165</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.008613</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Muribaculaceae  Bacteroides  Faecalibaculum  \\\n",
       "SampleID                                                 \n",
       "sample131        0.237934     0.004178        0.237536   \n",
       "sample132        0.217026     0.022603        0.148081   \n",
       "sample133        0.236170     0.013355        0.213967   \n",
       "sample134        0.181429     0.019500        0.116534   \n",
       "sample135        0.225473     0.011760        0.218809   \n",
       "\n",
       "           Bacteroides_dash_acidifaciens  Parabacteroides_dash_goldsteinii  \\\n",
       "SampleID                                                                     \n",
       "sample131                       0.018370                          0.001134   \n",
       "sample132                       0.038074                          0.000506   \n",
       "sample133                       0.022921                          0.000792   \n",
       "sample134                       0.048061                          0.000864   \n",
       "sample135                       0.053997                          0.000817   \n",
       "\n",
       "           Parasutterella  Akkermansia_dash_muciniphila  Lachnospiraceae  \\\n",
       "SampleID                                                                   \n",
       "sample131        0.024167                      0.041179         0.036914   \n",
       "sample132        0.030693                      0.022308         0.038992   \n",
       "sample133        0.081351                      0.063828         0.012851   \n",
       "sample134        0.023371                      0.106147         0.038334   \n",
       "sample135        0.035780                      0.053453         0.014144   \n",
       "\n",
       "           Alloprevotella  Lachnospiraceae_dash_NK4A136_dash_group  \\\n",
       "SampleID                                                             \n",
       "sample131        0.106118                                 0.029702   \n",
       "sample132        0.002435                                 0.031797   \n",
       "sample133        0.017804                                 0.012587   \n",
       "sample134        0.017331                                 0.019330   \n",
       "sample135        0.020514                                 0.012380   \n",
       "\n",
       "           Lachnospiraceae_dash_bacterium_dash_28_dash_4  Desulfovibrionaceae  \\\n",
       "SampleID                                                                        \n",
       "sample131                                       0.001105             0.001987   \n",
       "sample132                                       0.067491             0.008470   \n",
       "sample133                                       0.003566             0.002575   \n",
       "sample134                                       0.000735             0.010890   \n",
       "sample135                                       0.000686             0.003082   \n",
       "\n",
       "           Rikenellaceae_dash_RC9_dash_gut_dash_group  Oscillospiraceae  \\\n",
       "SampleID                                                                  \n",
       "sample131                                    0.074031          0.006824   \n",
       "sample132                                    0.029207          0.033602   \n",
       "sample133                                    0.086568          0.009055   \n",
       "sample134                                    0.108460          0.023867   \n",
       "sample135                                    0.120364          0.006152   \n",
       "\n",
       "           Clostridia_dash_vadinBB60_dash_group  Bacteroides_dash_uniformis  \\\n",
       "SampleID                                                                      \n",
       "sample131                              0.005138                    0.000756   \n",
       "sample132                              0.001221                    0.000856   \n",
       "sample133                              0.002996                    0.000850   \n",
       "sample134                              0.000911                    0.000592   \n",
       "sample135                              0.000512                    0.000283   \n",
       "\n",
       "           Clostridia_dash_UCG_dash_014  Alistipes  Helicobacter_dash_ganmani  \\\n",
       "SampleID                                                                        \n",
       "sample131                      0.001861   0.011623                   0.000058   \n",
       "sample132                      0.000614   0.010656                   0.000101   \n",
       "sample133                      0.000949   0.018225                   0.000091   \n",
       "sample134                      0.000326   0.037851                   0.000095   \n",
       "sample135                      0.006032   0.032165                   0.000076   \n",
       "\n",
       "           Mucispirillum_dash_schaedleri  \n",
       "SampleID                                  \n",
       "sample131                       0.021927  \n",
       "sample132                       0.114954  \n",
       "sample133                       0.069399  \n",
       "sample134                       0.023860  \n",
       "sample135                       0.008613  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topX = 20\n",
    "df_16S_T = df_16S.loc[df_meta.index].T\n",
    "df_16S_T['mean'] = df_16S_T.mean(axis=1)\n",
    "df_16S_T = df_16S_T.sort_values(by=['mean'],axis=0,ascending=False)\n",
    "df_16S_T = df_16S_T.drop('mean', axis=1)\n",
    "df_16S_topX = df_16S_T.iloc[0:topX].T\n",
    "df_16S_topX.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Muribaculaceae',\n",
       " 'Bacteroides',\n",
       " 'Faecalibaculum',\n",
       " 'Bacteroides_dash_acidifaciens',\n",
       " 'Parabacteroides_dash_goldsteinii',\n",
       " 'Parasutterella',\n",
       " 'Akkermansia_dash_muciniphila',\n",
       " 'Lachnospiraceae',\n",
       " 'Alloprevotella',\n",
       " 'Lachnospiraceae_dash_NK4A136_dash_group',\n",
       " 'Lachnospiraceae_dash_bacterium_dash_28_dash_4',\n",
       " 'Desulfovibrionaceae',\n",
       " 'Rikenellaceae_dash_RC9_dash_gut_dash_group',\n",
       " 'Oscillospiraceae',\n",
       " 'Clostridia_dash_vadinBB60_dash_group',\n",
       " 'Bacteroides_dash_uniformis',\n",
       " 'Clostridia_dash_UCG_dash_014',\n",
       " 'Alistipes',\n",
       " 'Helicobacter_dash_ganmani',\n",
       " 'Mucispirillum_dash_schaedleri']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_16S_topX.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize max value of 16S data to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MiceID</th>\n",
       "      <th>Vendor</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Muribaculaceae</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibaculum</th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Parabacteroides_dash_goldsteinii</th>\n",
       "      <th>Parasutterella</th>\n",
       "      <th>...</th>\n",
       "      <th>Lachnospiraceae_dash_bacterium_dash_28_dash_4</th>\n",
       "      <th>Desulfovibrionaceae</th>\n",
       "      <th>Rikenellaceae_dash_RC9_dash_gut_dash_group</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Clostridia_dash_vadinBB60_dash_group</th>\n",
       "      <th>Bacteroides_dash_uniformis</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Helicobacter_dash_ganmani</th>\n",
       "      <th>Mucispirillum_dash_schaedleri</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>BC1</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.318127</td>\n",
       "      <td>0.005586</td>\n",
       "      <td>0.317596</td>\n",
       "      <td>0.024561</td>\n",
       "      <td>0.001516</td>\n",
       "      <td>0.032312</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001478</td>\n",
       "      <td>0.002657</td>\n",
       "      <td>0.098983</td>\n",
       "      <td>0.009125</td>\n",
       "      <td>0.006869</td>\n",
       "      <td>0.001011</td>\n",
       "      <td>0.002489</td>\n",
       "      <td>0.015540</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.029318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>BC2</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.290174</td>\n",
       "      <td>0.030222</td>\n",
       "      <td>0.197990</td>\n",
       "      <td>0.050907</td>\n",
       "      <td>0.000676</td>\n",
       "      <td>0.041037</td>\n",
       "      <td>...</td>\n",
       "      <td>0.090239</td>\n",
       "      <td>0.011325</td>\n",
       "      <td>0.039051</td>\n",
       "      <td>0.044927</td>\n",
       "      <td>0.001633</td>\n",
       "      <td>0.001144</td>\n",
       "      <td>0.000822</td>\n",
       "      <td>0.014248</td>\n",
       "      <td>0.000135</td>\n",
       "      <td>0.153698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>BC3</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.315770</td>\n",
       "      <td>0.017856</td>\n",
       "      <td>0.286083</td>\n",
       "      <td>0.030647</td>\n",
       "      <td>0.001059</td>\n",
       "      <td>0.108770</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004768</td>\n",
       "      <td>0.003443</td>\n",
       "      <td>0.115744</td>\n",
       "      <td>0.012106</td>\n",
       "      <td>0.004006</td>\n",
       "      <td>0.001137</td>\n",
       "      <td>0.001269</td>\n",
       "      <td>0.024367</td>\n",
       "      <td>0.000121</td>\n",
       "      <td>0.092790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>BC4</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.242578</td>\n",
       "      <td>0.026073</td>\n",
       "      <td>0.155810</td>\n",
       "      <td>0.064259</td>\n",
       "      <td>0.001155</td>\n",
       "      <td>0.031248</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000982</td>\n",
       "      <td>0.014560</td>\n",
       "      <td>0.145016</td>\n",
       "      <td>0.031911</td>\n",
       "      <td>0.001219</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000437</td>\n",
       "      <td>0.050609</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.031902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>BC5</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.301467</td>\n",
       "      <td>0.015723</td>\n",
       "      <td>0.292557</td>\n",
       "      <td>0.072197</td>\n",
       "      <td>0.001092</td>\n",
       "      <td>0.047840</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000917</td>\n",
       "      <td>0.004120</td>\n",
       "      <td>0.160932</td>\n",
       "      <td>0.008226</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>0.000379</td>\n",
       "      <td>0.008066</td>\n",
       "      <td>0.043006</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>0.011516</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          MiceID   Vendor  Day  Diet  Muribaculaceae  Bacteroides  \\\n",
       "SampleID                                                            \n",
       "sample131    BC1  Beijing    0     0        0.318127     0.005586   \n",
       "sample132    BC2  Beijing    0     0        0.290174     0.030222   \n",
       "sample133    BC3  Beijing    0     0        0.315770     0.017856   \n",
       "sample134    BC4  Beijing    0     0        0.242578     0.026073   \n",
       "sample135    BC5  Beijing    0     0        0.301467     0.015723   \n",
       "\n",
       "           Faecalibaculum  Bacteroides_dash_acidifaciens  \\\n",
       "SampleID                                                   \n",
       "sample131        0.317596                       0.024561   \n",
       "sample132        0.197990                       0.050907   \n",
       "sample133        0.286083                       0.030647   \n",
       "sample134        0.155810                       0.064259   \n",
       "sample135        0.292557                       0.072197   \n",
       "\n",
       "           Parabacteroides_dash_goldsteinii  Parasutterella  ...  \\\n",
       "SampleID                                                     ...   \n",
       "sample131                          0.001516        0.032312  ...   \n",
       "sample132                          0.000676        0.041037  ...   \n",
       "sample133                          0.001059        0.108770  ...   \n",
       "sample134                          0.001155        0.031248  ...   \n",
       "sample135                          0.001092        0.047840  ...   \n",
       "\n",
       "           Lachnospiraceae_dash_bacterium_dash_28_dash_4  Desulfovibrionaceae  \\\n",
       "SampleID                                                                        \n",
       "sample131                                       0.001478             0.002657   \n",
       "sample132                                       0.090239             0.011325   \n",
       "sample133                                       0.004768             0.003443   \n",
       "sample134                                       0.000982             0.014560   \n",
       "sample135                                       0.000917             0.004120   \n",
       "\n",
       "           Rikenellaceae_dash_RC9_dash_gut_dash_group  Oscillospiraceae  \\\n",
       "SampleID                                                                  \n",
       "sample131                                    0.098983          0.009125   \n",
       "sample132                                    0.039051          0.044927   \n",
       "sample133                                    0.115744          0.012106   \n",
       "sample134                                    0.145016          0.031911   \n",
       "sample135                                    0.160932          0.008226   \n",
       "\n",
       "           Clostridia_dash_vadinBB60_dash_group  Bacteroides_dash_uniformis  \\\n",
       "SampleID                                                                      \n",
       "sample131                              0.006869                    0.001011   \n",
       "sample132                              0.001633                    0.001144   \n",
       "sample133                              0.004006                    0.001137   \n",
       "sample134                              0.001219                    0.000791   \n",
       "sample135                              0.000684                    0.000379   \n",
       "\n",
       "           Clostridia_dash_UCG_dash_014  Alistipes  Helicobacter_dash_ganmani  \\\n",
       "SampleID                                                                        \n",
       "sample131                      0.002489   0.015540                   0.000078   \n",
       "sample132                      0.000822   0.014248                   0.000135   \n",
       "sample133                      0.001269   0.024367                   0.000121   \n",
       "sample134                      0.000437   0.050609                   0.000127   \n",
       "sample135                      0.008066   0.043006                   0.000102   \n",
       "\n",
       "           Mucispirillum_dash_schaedleri  \n",
       "SampleID                                  \n",
       "sample131                       0.029318  \n",
       "sample132                       0.153698  \n",
       "sample133                       0.092790  \n",
       "sample134                       0.031902  \n",
       "sample135                       0.011516  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterial_taxa = list(df_16S_topX.columns)\n",
    "df_16S_topX_w_meta = pd.merge(df_meta, df_16S_topX/df_16S_topX.max().max(), left_index=True, right_index=True, how='inner')\n",
    "df_16S_topX_w_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate log-derivatives (dlog(B)/dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MiceID</th>\n",
       "      <th>Vendor</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Muribaculaceae</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibaculum</th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Parabacteroides_dash_goldsteinii</th>\n",
       "      <th>Parasutterella</th>\n",
       "      <th>...</th>\n",
       "      <th>Lachnospiraceae_dash_bacterium_dash_28_dash_4</th>\n",
       "      <th>Desulfovibrionaceae</th>\n",
       "      <th>Rikenellaceae_dash_RC9_dash_gut_dash_group</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Clostridia_dash_vadinBB60_dash_group</th>\n",
       "      <th>Bacteroides_dash_uniformis</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Helicobacter_dash_ganmani</th>\n",
       "      <th>Mucispirillum_dash_schaedleri</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>BC1</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.294894</td>\n",
       "      <td>0.198953</td>\n",
       "      <td>-0.687407</td>\n",
       "      <td>-0.019570</td>\n",
       "      <td>-0.000134</td>\n",
       "      <td>-0.080179</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.054459</td>\n",
       "      <td>0.364591</td>\n",
       "      <td>0.008018</td>\n",
       "      <td>0.030366</td>\n",
       "      <td>-0.004121</td>\n",
       "      <td>0.002133</td>\n",
       "      <td>0.008565</td>\n",
       "      <td>0.080135</td>\n",
       "      <td>0.015186</td>\n",
       "      <td>-0.030793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>BC2</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.369381</td>\n",
       "      <td>-0.022964</td>\n",
       "      <td>-0.527072</td>\n",
       "      <td>-0.083123</td>\n",
       "      <td>0.000630</td>\n",
       "      <td>-0.098706</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.134347</td>\n",
       "      <td>0.572466</td>\n",
       "      <td>-0.002834</td>\n",
       "      <td>-0.010603</td>\n",
       "      <td>0.004008</td>\n",
       "      <td>-0.001277</td>\n",
       "      <td>0.000674</td>\n",
       "      <td>0.002221</td>\n",
       "      <td>0.203150</td>\n",
       "      <td>-0.188570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>BC3</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.147410</td>\n",
       "      <td>0.191872</td>\n",
       "      <td>-0.541034</td>\n",
       "      <td>-0.073273</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>-0.236399</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004334</td>\n",
       "      <td>0.216482</td>\n",
       "      <td>-0.016506</td>\n",
       "      <td>0.047448</td>\n",
       "      <td>0.060661</td>\n",
       "      <td>-0.001317</td>\n",
       "      <td>0.023582</td>\n",
       "      <td>0.172864</td>\n",
       "      <td>0.056731</td>\n",
       "      <td>-0.137101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>BC4</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.151784</td>\n",
       "      <td>-0.006264</td>\n",
       "      <td>-0.149190</td>\n",
       "      <td>-0.075447</td>\n",
       "      <td>-0.000307</td>\n",
       "      <td>-0.013529</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005038</td>\n",
       "      <td>0.021675</td>\n",
       "      <td>0.026115</td>\n",
       "      <td>0.068254</td>\n",
       "      <td>0.016930</td>\n",
       "      <td>-0.001200</td>\n",
       "      <td>0.004031</td>\n",
       "      <td>-0.036927</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>-0.035878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>BC5</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.156434</td>\n",
       "      <td>0.079137</td>\n",
       "      <td>-0.378469</td>\n",
       "      <td>-0.050401</td>\n",
       "      <td>-0.000142</td>\n",
       "      <td>-0.030689</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017626</td>\n",
       "      <td>0.036313</td>\n",
       "      <td>-0.118671</td>\n",
       "      <td>0.015219</td>\n",
       "      <td>0.005966</td>\n",
       "      <td>0.000591</td>\n",
       "      <td>-0.007822</td>\n",
       "      <td>0.028085</td>\n",
       "      <td>-0.000087</td>\n",
       "      <td>-0.000541</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          MiceID   Vendor  Day  Diet  Muribaculaceae  Bacteroides  \\\n",
       "SampleID                                                            \n",
       "sample131    BC1  Beijing    0     0       -0.294894     0.198953   \n",
       "sample132    BC2  Beijing    0     0       -0.369381    -0.022964   \n",
       "sample133    BC3  Beijing    0     0       -0.147410     0.191872   \n",
       "sample134    BC4  Beijing    0     0        0.151784    -0.006264   \n",
       "sample135    BC5  Beijing    0     0        0.156434     0.079137   \n",
       "\n",
       "           Faecalibaculum  Bacteroides_dash_acidifaciens  \\\n",
       "SampleID                                                   \n",
       "sample131       -0.687407                      -0.019570   \n",
       "sample132       -0.527072                      -0.083123   \n",
       "sample133       -0.541034                      -0.073273   \n",
       "sample134       -0.149190                      -0.075447   \n",
       "sample135       -0.378469                      -0.050401   \n",
       "\n",
       "           Parabacteroides_dash_goldsteinii  Parasutterella  ...  \\\n",
       "SampleID                                                     ...   \n",
       "sample131                         -0.000134       -0.080179  ...   \n",
       "sample132                          0.000630       -0.098706  ...   \n",
       "sample133                          0.000083       -0.236399  ...   \n",
       "sample134                         -0.000307       -0.013529  ...   \n",
       "sample135                         -0.000142       -0.030689  ...   \n",
       "\n",
       "           Lachnospiraceae_dash_bacterium_dash_28_dash_4  Desulfovibrionaceae  \\\n",
       "SampleID                                                                        \n",
       "sample131                                      -0.054459             0.364591   \n",
       "sample132                                      -0.134347             0.572466   \n",
       "sample133                                      -0.004334             0.216482   \n",
       "sample134                                       0.005038             0.021675   \n",
       "sample135                                       0.017626             0.036313   \n",
       "\n",
       "           Rikenellaceae_dash_RC9_dash_gut_dash_group  Oscillospiraceae  \\\n",
       "SampleID                                                                  \n",
       "sample131                                    0.008018          0.030366   \n",
       "sample132                                   -0.002834         -0.010603   \n",
       "sample133                                   -0.016506          0.047448   \n",
       "sample134                                    0.026115          0.068254   \n",
       "sample135                                   -0.118671          0.015219   \n",
       "\n",
       "           Clostridia_dash_vadinBB60_dash_group  Bacteroides_dash_uniformis  \\\n",
       "SampleID                                                                      \n",
       "sample131                             -0.004121                    0.002133   \n",
       "sample132                              0.004008                   -0.001277   \n",
       "sample133                              0.060661                   -0.001317   \n",
       "sample134                              0.016930                   -0.001200   \n",
       "sample135                              0.005966                    0.000591   \n",
       "\n",
       "           Clostridia_dash_UCG_dash_014  Alistipes  Helicobacter_dash_ganmani  \\\n",
       "SampleID                                                                        \n",
       "sample131                      0.008565   0.080135                   0.015186   \n",
       "sample132                      0.000674   0.002221                   0.203150   \n",
       "sample133                      0.023582   0.172864                   0.056731   \n",
       "sample134                      0.004031  -0.036927                   0.000140   \n",
       "sample135                     -0.007822   0.028085                  -0.000087   \n",
       "\n",
       "           Mucispirillum_dash_schaedleri  \n",
       "SampleID                                  \n",
       "sample131                      -0.030793  \n",
       "sample132                      -0.188570  \n",
       "sample133                      -0.137101  \n",
       "sample134                      -0.035878  \n",
       "sample135                      -0.000541  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_deriv = deepcopy(df_16S_topX_w_meta)\n",
    "for curr_mice in set(df_deriv.MiceID):\n",
    "    curr_df = df_deriv[df_deriv.MiceID==curr_mice]\n",
    "    for taxon in bacterial_taxa:\n",
    "        xdata = np.array(curr_df['Day'])\n",
    "        ydata = np.array(curr_df[taxon])\n",
    "        cs = CubicSpline(xdata, ydata)\n",
    "        csd1 = cs.derivative(nu=1)\n",
    "        ydata_d1 = csd1(xdata)\n",
    "        df_deriv.loc[df_deriv.MiceID==curr_mice, taxon] = ydata_d1\n",
    "df_deriv.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct regression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-5.38285709e+00, -6.74087977e+00, -2.69396521e+00, ...,\n",
       "       -4.34048524e-02, -4.58747292e-04, -1.03068191e-02])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ymat should be samples by bacteria\n",
    "Ymat = df_deriv[bacterial_taxa].values\n",
    "Ymat = Ymat.flatten(order='F')\n",
    "Ymat = StandardScaler().fit_transform(Ymat.reshape(-1,1)).reshape(1,-1)[0] # standardize\n",
    "Ymat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6740"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Ymat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.00000000e+00, 0.00000000e+00, 3.18127455e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 0.00000000e+00, 2.90173547e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 0.00000000e+00, 3.15770007e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       ...,\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        1.39557067e-04, 1.61027385e-04, 8.91018196e-04],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        2.57211509e-04, 2.47685157e-04, 8.19266287e-04],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        7.54193953e-05, 1.43296851e-04, 1.13129093e-04]])"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Xmat = np.zeros(shape=(topX*len(df_deriv.index), (topX+2)*topX))\n",
    "for k in np.arange(topX):\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)] = 1\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+1] = df_deriv.Diet.values\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+2:(k+1)*(topX+2)] = df_16S_topX_w_meta[bacterial_taxa].values\n",
    "Xmat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write stan program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "json_str = '{\\n\"N\" : %d,\\n'%(len(Ymat))\n",
    "\n",
    "# y variable\n",
    "json_str += '\\\"dlogX\\\" : [%s],\\n'%(','.join(list(Ymat.astype(str))))\n",
    "\n",
    "# x variable\n",
    "for k1,c1 in enumerate(bacterial_taxa):\n",
    "    # growth rate\n",
    "    json_str += '\\\"growth_rate_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)].astype(str))))\n",
    "    # diet response\n",
    "    json_str += '\\\"inulin_response_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)+1].astype(str))))\n",
    "    # bacterial interactions\n",
    "    for k2,c2 in enumerate(bacterial_taxa):\n",
    "        v = list(Xmat[:,k1*(topX+2)+2+k2].astype(str))\n",
    "        json_str += '\\\"pairwise_interaction_%s_%s\\\" : [%s]'%(c1,c2,','.join(v))\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            json_str += '\\n}'\n",
    "        else:\n",
    "            json_str += ',\\n'\n",
    "#print(json_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.data.json\", \"w\")\n",
    "text_file.write(\"%s\" % json_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# data block\n",
    "model_str = 'data {\\n'\n",
    "model_str += '\\tint<lower=0> N;\\n'\n",
    "model_str += '\\tvector[N] dlogX;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\tvector[N] growth_rate_%s;\\n'%(c1)\n",
    "    model_str += '\\tvector[N] inulin_response_%s;\\n'%(c1)\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tvector[N] pairwise_interaction_%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'\n",
    "\n",
    "# parameter block\n",
    "model_str += 'parameters {\\n\\treal<lower=0,upper=1> sigma;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\treal alpha__%s;\\n'%(c1) # growth rate\n",
    "    model_str += '\\treal epsilon__%s;\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\treal beta__%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'       \n",
    "        \n",
    "# model block\n",
    "model_str += 'model {\\n\\tsigma ~ uniform(0,1);\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\talpha__%s ~ normal(0,1);\\n'%(c1) # growth rate\n",
    "    model_str += '\\tepsilon__%s ~ normal(0,1);\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tbeta__%s_%s ~ normal(0,1);\\n'%(c1,c2)\n",
    "model_str += '\\tdlogX ~ normal('\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += 'alpha__%s*growth_rate_%s+'%(c1,c1) # growth rate\n",
    "    model_str += 'epsilon__%s*inulin_response_%s+'%(c1,c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s'%(c1,c2,c1,c2)\n",
    "        else:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s+'%(c1,c2,c1,c2)\n",
    "model_str += ', sigma);\\n}'\n",
    "#print(model_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.stan\", \"w\")\n",
    "text_file.write(\"%s\" % model_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot stan output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hpd_grid(sample, alpha=0.05, roundto=2):\n",
    "    \"\"\"Calculate highest posterior density (HPD) of array for given alpha. \n",
    "    The HPD is the minimum width Bayesian credible interval (BCI). \n",
    "    The function works for multimodal distributions, returning more than one mode\n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    sample : Numpy array or python list\n",
    "        An array containing MCMC samples\n",
    "    alpha : float\n",
    "        Desired probability of type I error (defaults to 0.05)\n",
    "    roundto: integer\n",
    "        Number of digits after the decimal point for the results\n",
    "    Returns\n",
    "    ----------\n",
    "    hpd: array with the lower \n",
    "          \n",
    "    \"\"\"\n",
    "    sample = np.asarray(sample)\n",
    "    sample = sample[~np.isnan(sample)]\n",
    "    # get upper and lower bounds\n",
    "    l = np.min(sample)\n",
    "    u = np.max(sample)\n",
    "    density = kde.gaussian_kde(sample)\n",
    "    x = np.linspace(l, u, 2000)\n",
    "    y = density.evaluate(x)\n",
    "    #y = density.evaluate(x, l, u) waitting for PR to be accepted\n",
    "    xy_zipped = zip(x, y/np.sum(y))\n",
    "    xy = sorted(xy_zipped, key=lambda x: x[1], reverse=True)\n",
    "    xy_cum_sum = 0\n",
    "    hdv = []\n",
    "    for val in xy:\n",
    "        xy_cum_sum += val[1]\n",
    "        hdv.append(val[0])\n",
    "        if xy_cum_sum >= (1-alpha):\n",
    "            break\n",
    "    hdv.sort()\n",
    "    diff = (u-l)/20  # differences of 5%\n",
    "    hpd = []\n",
    "    hpd.append(round(min(hdv), roundto))\n",
    "    for i in range(1, len(hdv)):\n",
    "        if hdv[i]-hdv[i-1] >= diff:\n",
    "            hpd.append(round(hdv[i-1], roundto))\n",
    "            hpd.append(round(hdv[i], roundto))\n",
    "    hpd.append(round(max(hdv), roundto))\n",
    "    ite = iter(hpd)\n",
    "    hpd = list(zip(ite, ite))\n",
    "    modes = []\n",
    "    for value in hpd:\n",
    "        x_hpd = x[(x > value[0]) & (x < value[1])]\n",
    "        y_hpd = y[(x > value[0]) & (x < value[1])]\n",
    "        modes.append(round(x_hpd[np.argmax(y_hpd)], roundto))\n",
    "    return hpd, x, y, modes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = az.from_cmdstan([\"output_%d.csv\"%(i) for i in np.arange(1,4)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Muribaculaceae</td>\n",
       "      <td>-0.16</td>\n",
       "      <td>0.15</td>\n",
       "      <td>-0.005</td>\n",
       "      <td>-0.044859</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bacteroides_dash_acidifaciens</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.145</td>\n",
       "      <td>1.835619</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Akkermansia_dash_muciniphila</td>\n",
       "      <td>-0.27</td>\n",
       "      <td>0.04</td>\n",
       "      <td>-0.115</td>\n",
       "      <td>-1.416896</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Faecalibaculum</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-0.040</td>\n",
       "      <td>-0.531562</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Parasutterella</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-0.040</td>\n",
       "      <td>-0.540205</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Type                           Taxa  Left  Right  Middle  \\\n",
       "0  basal_growth_rate                 Muribaculaceae -0.16   0.15  -0.005   \n",
       "1  basal_growth_rate  Bacteroides_dash_acidifaciens -0.01   0.30   0.145   \n",
       "2  basal_growth_rate   Akkermansia_dash_muciniphila -0.27   0.04  -0.115   \n",
       "3  basal_growth_rate                 Faecalibaculum -0.20   0.12  -0.040   \n",
       "4  basal_growth_rate                 Parasutterella -0.20   0.12  -0.040   \n",
       "\n",
       "        SNR  Significant  \n",
       "0 -0.044859        False  \n",
       "1  1.835619        False  \n",
       "2 -1.416896        False  \n",
       "3 -0.531562        False  \n",
       "4 -0.540205        False  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines = []\n",
    "    \n",
    "# basal growth rate\n",
    "for c in bacterial_taxa:\n",
    "    var = 'alpha__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['basal_growth_rate', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# inulin response\n",
    "for c in bacterial_taxa:\n",
    "    var = 'epsilon__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['inulin_response', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "\n",
    "# pairwise interactions\n",
    "for c1 in bacterial_taxa:\n",
    "    for c2 in bacterial_taxa:\n",
    "        var = 'beta__%s_%s'%(c1,c2)\n",
    "        data = []\n",
    "        for i in np.arange(0,3):\n",
    "            data.extend(list(fit.posterior[var][i].values))\n",
    "        hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "        assert len(hpd_mu) == 1\n",
    "        (x0, x1) = hpd_mu[0]\n",
    "        lines.append(['pairwise_interaction', (c1,c2), x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "df_stan_output_summary = pd.DataFrame(lines, columns = ['Type','Taxa','Left','Right','Middle','SNR','Significant'])\n",
    "df_stan_output_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stan_output_summary.to_excel('bayesian_regression_summary_top20.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Muribaculaceae</td>\n",
       "      <td>1.08</td>\n",
       "      <td>1.62</td>\n",
       "      <td>1.350</td>\n",
       "      <td>9.958040</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Bacteroides_dash_acidifaciens</td>\n",
       "      <td>1.89</td>\n",
       "      <td>2.42</td>\n",
       "      <td>2.155</td>\n",
       "      <td>15.846923</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Faecalibaculum</td>\n",
       "      <td>0.48</td>\n",
       "      <td>1.01</td>\n",
       "      <td>0.745</td>\n",
       "      <td>5.506163</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Parasutterella</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.410</td>\n",
       "      <td>3.039780</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.310</td>\n",
       "      <td>2.265775</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Muribaculaceae', 'Muribaculaceae')</td>\n",
       "      <td>-1.59</td>\n",
       "      <td>-0.29</td>\n",
       "      <td>-0.940</td>\n",
       "      <td>-2.864951</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Muribaculaceae', 'Bacteroides_dash_acidifaci...</td>\n",
       "      <td>0.23</td>\n",
       "      <td>2.31</td>\n",
       "      <td>1.270</td>\n",
       "      <td>2.392931</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Muribaculaceae', 'Desulfovibrionaceae')</td>\n",
       "      <td>0.27</td>\n",
       "      <td>3.71</td>\n",
       "      <td>1.990</td>\n",
       "      <td>2.280134</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Muribaculaceae', 'Clostridia_dash_UCG_dash_0...</td>\n",
       "      <td>-3.74</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>-1.900</td>\n",
       "      <td>-2.016358</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Muribaculac...</td>\n",
       "      <td>-3.81</td>\n",
       "      <td>-2.50</td>\n",
       "      <td>-3.155</td>\n",
       "      <td>-9.603366</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Akkermansia...</td>\n",
       "      <td>-4.57</td>\n",
       "      <td>-1.77</td>\n",
       "      <td>-3.170</td>\n",
       "      <td>-4.472396</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Faecalibacu...</td>\n",
       "      <td>-4.31</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>-2.655</td>\n",
       "      <td>-3.148145</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Parasuttere...</td>\n",
       "      <td>-4.26</td>\n",
       "      <td>-0.57</td>\n",
       "      <td>-2.415</td>\n",
       "      <td>-2.600989</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Bacteroides')</td>\n",
       "      <td>-5.89</td>\n",
       "      <td>-2.35</td>\n",
       "      <td>-4.120</td>\n",
       "      <td>-4.574359</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Akkermansia_dash_muciniphila', 'Bacteroides_...</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2.09</td>\n",
       "      <td>1.050</td>\n",
       "      <td>1.970520</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Faecalibaculum', 'Muribaculaceae')</td>\n",
       "      <td>-1.77</td>\n",
       "      <td>-0.47</td>\n",
       "      <td>-1.120</td>\n",
       "      <td>-3.398489</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Escherichia_dash_Shigella', 'Escherichia_das...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>3.69</td>\n",
       "      <td>2.145</td>\n",
       "      <td>2.731059</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Type                                               Taxa  \\\n",
       "20        inulin_response                                     Muribaculaceae   \n",
       "21        inulin_response                      Bacteroides_dash_acidifaciens   \n",
       "23        inulin_response                                     Faecalibaculum   \n",
       "24        inulin_response                                     Parasutterella   \n",
       "25        inulin_response                                        Bacteroides   \n",
       "40   pairwise_interaction               ('Muribaculaceae', 'Muribaculaceae')   \n",
       "41   pairwise_interaction  ('Muribaculaceae', 'Bacteroides_dash_acidifaci...   \n",
       "51   pairwise_interaction          ('Muribaculaceae', 'Desulfovibrionaceae')   \n",
       "52   pairwise_interaction  ('Muribaculaceae', 'Clostridia_dash_UCG_dash_0...   \n",
       "60   pairwise_interaction  ('Bacteroides_dash_acidifaciens', 'Muribaculac...   \n",
       "62   pairwise_interaction  ('Bacteroides_dash_acidifaciens', 'Akkermansia...   \n",
       "63   pairwise_interaction  ('Bacteroides_dash_acidifaciens', 'Faecalibacu...   \n",
       "64   pairwise_interaction  ('Bacteroides_dash_acidifaciens', 'Parasuttere...   \n",
       "65   pairwise_interaction   ('Bacteroides_dash_acidifaciens', 'Bacteroides')   \n",
       "81   pairwise_interaction  ('Akkermansia_dash_muciniphila', 'Bacteroides_...   \n",
       "100  pairwise_interaction               ('Faecalibaculum', 'Muribaculaceae')   \n",
       "376  pairwise_interaction  ('Escherichia_dash_Shigella', 'Escherichia_das...   \n",
       "\n",
       "     Left  Right  Middle        SNR  Significant  \n",
       "20   1.08   1.62   1.350   9.958040         True  \n",
       "21   1.89   2.42   2.155  15.846923         True  \n",
       "23   0.48   1.01   0.745   5.506163         True  \n",
       "24   0.14   0.68   0.410   3.039780         True  \n",
       "25   0.04   0.58   0.310   2.265775         True  \n",
       "40  -1.59  -0.29  -0.940  -2.864951         True  \n",
       "41   0.23   2.31   1.270   2.392931         True  \n",
       "51   0.27   3.71   1.990   2.280134         True  \n",
       "52  -3.74  -0.06  -1.900  -2.016358         True  \n",
       "60  -3.81  -2.50  -3.155  -9.603366         True  \n",
       "62  -4.57  -1.77  -3.170  -4.472396         True  \n",
       "63  -4.31  -1.00  -2.655  -3.148145         True  \n",
       "64  -4.26  -0.57  -2.415  -2.600989         True  \n",
       "65  -5.89  -2.35  -4.120  -4.574359         True  \n",
       "81   0.01   2.09   1.050   1.970520         True  \n",
       "100 -1.77  -0.47  -1.120  -3.398489         True  \n",
       "376  0.60   3.69   2.145   2.731059         True  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stan_output_summary = pd.read_excel('bayesian_regression_summary_top20.xlsx', index_col=0)\n",
    "df_stan_output_summary = df_stan_output_summary[df_stan_output_summary.Significant==True]\n",
    "df_stan_output_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
