{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from copy import deepcopy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from scipy.interpolate import CubicSpline\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import warnings\n",
    "import itertools\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "import statsmodels.stats.api as sms\n",
    "import scipy.stats.kde as kde"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read 16S data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2</th>\n",
       "      <th>ASF356</th>\n",
       "      <th>Acetatifactor</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Alloprevotella</th>\n",
       "      <th>Anaerotruncus</th>\n",
       "      <th>Aquimarina</th>\n",
       "      <th>Aquimarina_dash_muelleri</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Clostridium_rightsquarebracket__dash_cocleatum</th>\n",
       "      <th>_leftsquarebracket_Clostridium_rightsquarebracket__dash_innocuum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_brachy_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_coprostanoligenes_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_fissicatena_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_nodatum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>mouse_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003163</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007693</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.153254</td>\n",
       "      <td>0.031291</td>\n",
       "      <td>0.031291</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample100</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005143</td>\n",
       "      <td>0.012320</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003843</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample101</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006024</td>\n",
       "      <td>0.003967</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample102</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001437</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.025687</td>\n",
       "      <td>0.011724</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 116 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            A2    ASF356  Acetatifactor  Acinetobacter  \\\n",
       "SampleID                                                 \n",
       "sample1    0.0  0.000000       0.000000            0.0   \n",
       "sample10   0.0  0.000000       0.007693            0.0   \n",
       "sample100  0.0  0.000000       0.000000            0.0   \n",
       "sample101  0.0  0.000309       0.000000            0.0   \n",
       "sample102  0.0  0.001437       0.000000            0.0   \n",
       "\n",
       "           Akkermansia_dash_muciniphila  Alistipes  Alloprevotella  \\\n",
       "SampleID                                                             \n",
       "sample1                             0.0   0.000000        0.003163   \n",
       "sample10                            0.0   0.153254        0.031291   \n",
       "sample100                           0.0   0.005143        0.012320   \n",
       "sample101                           0.0   0.006024        0.003967   \n",
       "sample102                           0.0   0.025687        0.011724   \n",
       "\n",
       "           Anaerotruncus  Aquimarina  Aquimarina_dash_muelleri  ...  \\\n",
       "SampleID                                                        ...   \n",
       "sample1         0.000000         0.0                       0.0  ...   \n",
       "sample10        0.031291         0.0                       0.0  ...   \n",
       "sample100       0.000000         0.0                       0.0  ...   \n",
       "sample101       0.000000         0.0                       0.0  ...   \n",
       "sample102       0.000000         0.0                       0.0  ...   \n",
       "\n",
       "           _leftsquarebracket_Clostridium_rightsquarebracket__dash_cocleatum  \\\n",
       "SampleID                                                                       \n",
       "sample1                                                  0.0                   \n",
       "sample10                                                 0.0                   \n",
       "sample100                                                0.0                   \n",
       "sample101                                                0.0                   \n",
       "sample102                                                0.0                   \n",
       "\n",
       "           _leftsquarebracket_Clostridium_rightsquarebracket__dash_innocuum_dash_group  \\\n",
       "SampleID                                                                                 \n",
       "sample1                                                  0.0                             \n",
       "sample10                                                 0.0                             \n",
       "sample100                                                0.0                             \n",
       "sample101                                                0.0                             \n",
       "sample102                                                0.0                             \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_brachy_dash_group  \\\n",
       "SampleID                                                                               \n",
       "sample1                                                  0.0                           \n",
       "sample10                                                 0.0                           \n",
       "sample100                                                0.0                           \n",
       "sample101                                                0.0                           \n",
       "sample102                                                0.0                           \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_coprostanoligenes_dash_group  \\\n",
       "SampleID                                                                                          \n",
       "sample1                                                  0.0                                      \n",
       "sample10                                                 0.0                                      \n",
       "sample100                                                0.0                                      \n",
       "sample101                                                0.0                                      \n",
       "sample102                                                0.0                                      \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_fissicatena_dash_group  \\\n",
       "SampleID                                                                                    \n",
       "sample1                                                  0.0                                \n",
       "sample10                                                 0.0                                \n",
       "sample100                                                0.0                                \n",
       "sample101                                                0.0                                \n",
       "sample102                                                0.0                                \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_nodatum_dash_group  \\\n",
       "SampleID                                                                                \n",
       "sample1                                                  0.0                            \n",
       "sample10                                                 0.0                            \n",
       "sample100                                                0.0                            \n",
       "sample101                                                0.0                            \n",
       "sample102                                                0.0                            \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group  \\\n",
       "SampleID                                                                                     \n",
       "sample1                                                  0.0                                 \n",
       "sample10                                                 0.0                                 \n",
       "sample100                                                0.0                                 \n",
       "sample101                                                0.0                                 \n",
       "sample102                                                0.0                                 \n",
       "\n",
       "           gut_dash_metagenome  metagenome  mouse_dash_gut_dash_metagenome  \n",
       "SampleID                                                                    \n",
       "sample1               0.000000         0.0                        0.000000  \n",
       "sample10              0.000000         0.0                        0.000000  \n",
       "sample100             0.003843         0.0                        0.000000  \n",
       "sample101             0.000000         0.0                        0.001019  \n",
       "sample102             0.003244         0.0                        0.000000  \n",
       "\n",
       "[5 rows x 116 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_16S = pd.read_excel('./16S_relative_abundance.xlsx', index_col=0)\n",
    "df_16S.columns = [c.replace('/','_slash_').replace(' ','_space_').replace('[','_leftsquarebracket_').replace(']','_rightsquarebracket_').replace('-','_dash_').replace('.','_dot_').replace('(','_leftroundbracket').replace(')','_rightroundbracket_') for c in df_16S.columns]\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add pseudocount for zeros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2</th>\n",
       "      <th>ASF356</th>\n",
       "      <th>Acetatifactor</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Alloprevotella</th>\n",
       "      <th>Anaerotruncus</th>\n",
       "      <th>Aquimarina</th>\n",
       "      <th>Aquimarina_dash_muelleri</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Clostridium_rightsquarebracket__dash_cocleatum</th>\n",
       "      <th>_leftsquarebracket_Clostridium_rightsquarebracket__dash_innocuum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_brachy_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_coprostanoligenes_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_fissicatena_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_nodatum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>mouse_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.003163</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample10</th>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.007693</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.153254</td>\n",
       "      <td>0.031291</td>\n",
       "      <td>0.031291</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.000432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample100</th>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.005143</td>\n",
       "      <td>0.012320</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.003843</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample101</th>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.006024</td>\n",
       "      <td>0.003967</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.001019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample102</th>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.001437</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.025687</td>\n",
       "      <td>0.011724</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.003244</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000082</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 116 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 A2    ASF356  Acetatifactor  Acinetobacter  \\\n",
       "SampleID                                                      \n",
       "sample1    0.000791  0.000791       0.000791       0.000791   \n",
       "sample10   0.000432  0.000432       0.007693       0.000432   \n",
       "sample100  0.000085  0.000085       0.000085       0.000085   \n",
       "sample101  0.000055  0.000309       0.000055       0.000055   \n",
       "sample102  0.000082  0.001437       0.000082       0.000082   \n",
       "\n",
       "           Akkermansia_dash_muciniphila  Alistipes  Alloprevotella  \\\n",
       "SampleID                                                             \n",
       "sample1                        0.000791   0.000791        0.003163   \n",
       "sample10                       0.000432   0.153254        0.031291   \n",
       "sample100                      0.000085   0.005143        0.012320   \n",
       "sample101                      0.000055   0.006024        0.003967   \n",
       "sample102                      0.000082   0.025687        0.011724   \n",
       "\n",
       "           Anaerotruncus  Aquimarina  Aquimarina_dash_muelleri  ...  \\\n",
       "SampleID                                                        ...   \n",
       "sample1         0.000791    0.000791                  0.000791  ...   \n",
       "sample10        0.031291    0.000432                  0.000432  ...   \n",
       "sample100       0.000085    0.000085                  0.000085  ...   \n",
       "sample101       0.000055    0.000055                  0.000055  ...   \n",
       "sample102       0.000082    0.000082                  0.000082  ...   \n",
       "\n",
       "           _leftsquarebracket_Clostridium_rightsquarebracket__dash_cocleatum  \\\n",
       "SampleID                                                                       \n",
       "sample1                                             0.000791                   \n",
       "sample10                                            0.000432                   \n",
       "sample100                                           0.000085                   \n",
       "sample101                                           0.000055                   \n",
       "sample102                                           0.000082                   \n",
       "\n",
       "           _leftsquarebracket_Clostridium_rightsquarebracket__dash_innocuum_dash_group  \\\n",
       "SampleID                                                                                 \n",
       "sample1                                             0.000791                             \n",
       "sample10                                            0.000432                             \n",
       "sample100                                           0.000085                             \n",
       "sample101                                           0.000055                             \n",
       "sample102                                           0.000082                             \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_brachy_dash_group  \\\n",
       "SampleID                                                                               \n",
       "sample1                                             0.000791                           \n",
       "sample10                                            0.000432                           \n",
       "sample100                                           0.000085                           \n",
       "sample101                                           0.000055                           \n",
       "sample102                                           0.000082                           \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_coprostanoligenes_dash_group  \\\n",
       "SampleID                                                                                          \n",
       "sample1                                             0.000791                                      \n",
       "sample10                                            0.000432                                      \n",
       "sample100                                           0.000085                                      \n",
       "sample101                                           0.000055                                      \n",
       "sample102                                           0.000082                                      \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_fissicatena_dash_group  \\\n",
       "SampleID                                                                                    \n",
       "sample1                                             0.000791                                \n",
       "sample10                                            0.000432                                \n",
       "sample100                                           0.000085                                \n",
       "sample101                                           0.000055                                \n",
       "sample102                                           0.000082                                \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_nodatum_dash_group  \\\n",
       "SampleID                                                                                \n",
       "sample1                                             0.000791                            \n",
       "sample10                                            0.000432                            \n",
       "sample100                                           0.000085                            \n",
       "sample101                                           0.000055                            \n",
       "sample102                                           0.000082                            \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group  \\\n",
       "SampleID                                                                                     \n",
       "sample1                                             0.000791                                 \n",
       "sample10                                            0.000432                                 \n",
       "sample100                                           0.000085                                 \n",
       "sample101                                           0.000055                                 \n",
       "sample102                                           0.000082                                 \n",
       "\n",
       "           gut_dash_metagenome  metagenome  mouse_dash_gut_dash_metagenome  \n",
       "SampleID                                                                    \n",
       "sample1               0.000791    0.000791                        0.000791  \n",
       "sample10              0.000432    0.000432                        0.000432  \n",
       "sample100             0.003843    0.000085                        0.000085  \n",
       "sample101             0.000055    0.000055                        0.001019  \n",
       "sample102             0.003244    0.000082                        0.000082  \n",
       "\n",
       "[5 rows x 116 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for sample_id in df_16S.index:\n",
    "    sample = np.array(df_16S.loc[sample_id])\n",
    "    minval = np.min(sample[np.nonzero(sample)]) # minimum non-zero value\n",
    "    sample[sample==0] = minval\n",
    "    df_16S.loc[sample_id] = sample\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read metadata for inulin and control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mice_ID</th>\n",
       "      <th>Day</th>\n",
       "      <th>Group</th>\n",
       "      <th>Diet</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>1000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample2</th>\n",
       "      <td>1000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample3</th>\n",
       "      <td>1000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample4</th>\n",
       "      <td>1000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample5</th>\n",
       "      <td>1000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Mice_ID  Day      Group  Diet\n",
       "#SampleID                               \n",
       "sample1    1000000  0.0  Cellulose     0\n",
       "sample2    1000000  0.0  Cellulose     0\n",
       "sample3    1000000  0.0  Cellulose     0\n",
       "sample4    1000000  0.0  Cellulose     0\n",
       "sample5    1000000  0.0  Cellulose     0"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv('metadata.txt', sep='\\t').set_index('#SampleID')\n",
    "df_meta = df_meta.drop('#q2:types')\n",
    "\n",
    "# remove inulin in non-longitudinal study\n",
    "df_meta = df_meta[~((df_meta.If_longitudinal=='No') & (df_meta.Group=='Inulin'))]\n",
    "\n",
    "# correct days: add 0.5 day to evening samples\n",
    "new_days = []\n",
    "for d,t in zip(df_meta.Day,df_meta.Time):\n",
    "    if t=='Evening':\n",
    "        new_days.append(float(d)+0.5)\n",
    "    else:\n",
    "        new_days.append(float(d))\n",
    "df_meta['Day'] = new_days\n",
    "\n",
    "# Correct for mice ID\n",
    "new_miceid = []\n",
    "for m in df_meta.Mice_ID:\n",
    "    if '_' in m:\n",
    "        new_miceid.append(1000000)\n",
    "    else:\n",
    "        new_miceid.append(m)\n",
    "df_meta['Mice_ID'] = new_miceid\n",
    "\n",
    "# change before to cellulose\n",
    "df_meta.loc[df_meta.Group=='Before', 'Group'] = 'Cellulose'\n",
    "        \n",
    "df_meta = df_meta[['Mice_ID','Day','Group']]\n",
    "df_meta['Diet'] = [1 if g=='Inulin' else 0 for g in df_meta['Group']]\n",
    "\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample average"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mice_ID</th>\n",
       "      <th>Day</th>\n",
       "      <th>Group</th>\n",
       "      <th>Diet</th>\n",
       "      <th>A2</th>\n",
       "      <th>ASF356</th>\n",
       "      <th>Acetatifactor</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Clostridium_rightsquarebracket__dash_cocleatum</th>\n",
       "      <th>_leftsquarebracket_Clostridium_rightsquarebracket__dash_innocuum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_brachy_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_coprostanoligenes_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_fissicatena_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_nodatum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>mouse_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.002132</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.013110</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.001798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000669</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.002674</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.101401</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000537</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000000</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000600</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.001202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1000000</td>\n",
       "      <td>14.5</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.008515</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.030853</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.002369</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Inulin</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000266</td>\n",
       "      <td>0.000349</td>\n",
       "      <td>0.001081</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.036614</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.002529</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.002578</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>0.045397</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000491</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 120 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mice_ID   Day      Group  Diet        A2    ASF356  Acetatifactor  \\\n",
       "0  1000000   0.0  Cellulose     0  0.000696  0.000696       0.002132   \n",
       "1  1000000   0.5  Cellulose     0  0.000669  0.000513       0.002674   \n",
       "2  1000000  14.0  Cellulose     0  0.000600  0.000536       0.000536   \n",
       "3  1000000  14.5  Cellulose     0  0.001340  0.001340       0.008515   \n",
       "4        1   0.0     Inulin     1  0.000266  0.000349       0.001081   \n",
       "\n",
       "   Acinetobacter  Akkermansia_dash_muciniphila  Alistipes  ...  \\\n",
       "0       0.000696                      0.000696   0.013110  ...   \n",
       "1       0.000513                      0.000513   0.101401  ...   \n",
       "2       0.000536                      0.000536   0.000536  ...   \n",
       "3       0.001340                      0.001340   0.030853  ...   \n",
       "4       0.000017                      0.000017   0.036614  ...   \n",
       "\n",
       "   _leftsquarebracket_Clostridium_rightsquarebracket__dash_cocleatum  \\\n",
       "0                                           0.000696                   \n",
       "1                                           0.000513                   \n",
       "2                                           0.000536                   \n",
       "3                                           0.001340                   \n",
       "4                                           0.000017                   \n",
       "\n",
       "   _leftsquarebracket_Clostridium_rightsquarebracket__dash_innocuum_dash_group  \\\n",
       "0                                           0.000696                             \n",
       "1                                           0.000513                             \n",
       "2                                           0.000536                             \n",
       "3                                           0.001340                             \n",
       "4                                           0.000166                             \n",
       "\n",
       "   _leftsquarebracket_Eubacterium_rightsquarebracket__dash_brachy_dash_group  \\\n",
       "0                                           0.000696                           \n",
       "1                                           0.000513                           \n",
       "2                                           0.000536                           \n",
       "3                                           0.001340                           \n",
       "4                                           0.002529                           \n",
       "\n",
       "   _leftsquarebracket_Eubacterium_rightsquarebracket__dash_coprostanoligenes_dash_group  \\\n",
       "0                                           0.000696                                      \n",
       "1                                           0.000537                                      \n",
       "2                                           0.000536                                      \n",
       "3                                           0.001340                                      \n",
       "4                                           0.000017                                      \n",
       "\n",
       "   _leftsquarebracket_Eubacterium_rightsquarebracket__dash_fissicatena_dash_group  \\\n",
       "0                                           0.000696                                \n",
       "1                                           0.000513                                \n",
       "2                                           0.000536                                \n",
       "3                                           0.002369                                \n",
       "4                                           0.000017                                \n",
       "\n",
       "   _leftsquarebracket_Eubacterium_rightsquarebracket__dash_nodatum_dash_group  \\\n",
       "0                                           0.000696                            \n",
       "1                                           0.000513                            \n",
       "2                                           0.000536                            \n",
       "3                                           0.001340                            \n",
       "4                                           0.002578                            \n",
       "\n",
       "   _leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group  \\\n",
       "0                                           0.000696                                 \n",
       "1                                           0.000513                                 \n",
       "2                                           0.000536                                 \n",
       "3                                           0.001340                                 \n",
       "4                                           0.000183                                 \n",
       "\n",
       "   gut_dash_metagenome  metagenome  mouse_dash_gut_dash_metagenome  \n",
       "0             0.000696    0.000696                        0.001798  \n",
       "1             0.000513    0.000513                        0.000513  \n",
       "2             0.000536    0.000536                        0.001202  \n",
       "3             0.001340    0.001340                        0.001815  \n",
       "4             0.045397    0.000017                        0.000491  \n",
       "\n",
       "[5 rows x 120 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_16S_meta = pd.merge(df_meta, df_16S, left_index=True, right_index=True, how='inner')\n",
    "df_16S_meta = df_16S_meta.groupby(['Mice_ID','Day','Group','Diet']).agg(np.mean).reset_index()\n",
    "df_16S_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2</th>\n",
       "      <th>ASF356</th>\n",
       "      <th>Acetatifactor</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Alloprevotella</th>\n",
       "      <th>Anaerotruncus</th>\n",
       "      <th>Aquimarina</th>\n",
       "      <th>Aquimarina_dash_muelleri</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Clostridium_rightsquarebracket__dash_cocleatum</th>\n",
       "      <th>_leftsquarebracket_Clostridium_rightsquarebracket__dash_innocuum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_brachy_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_coprostanoligenes_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_fissicatena_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_nodatum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>mouse_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.002132</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.013110</td>\n",
       "      <td>0.010021</td>\n",
       "      <td>0.005460</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.001798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000669</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.002674</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.101401</td>\n",
       "      <td>0.022954</td>\n",
       "      <td>0.014820</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000537</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000600</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.003964</td>\n",
       "      <td>0.000557</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.001202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.008515</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.030853</td>\n",
       "      <td>0.014149</td>\n",
       "      <td>0.002125</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.002369</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000266</td>\n",
       "      <td>0.000349</td>\n",
       "      <td>0.001081</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.036614</td>\n",
       "      <td>0.005797</td>\n",
       "      <td>0.008742</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.002529</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.002578</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>0.045397</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000491</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 116 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         A2    ASF356  Acetatifactor  Acinetobacter  \\\n",
       "0  0.000696  0.000696       0.002132       0.000696   \n",
       "1  0.000669  0.000513       0.002674       0.000513   \n",
       "2  0.000600  0.000536       0.000536       0.000536   \n",
       "3  0.001340  0.001340       0.008515       0.001340   \n",
       "4  0.000266  0.000349       0.001081       0.000017   \n",
       "\n",
       "   Akkermansia_dash_muciniphila  Alistipes  Alloprevotella  Anaerotruncus  \\\n",
       "0                      0.000696   0.013110        0.010021       0.005460   \n",
       "1                      0.000513   0.101401        0.022954       0.014820   \n",
       "2                      0.000536   0.000536        0.003964       0.000557   \n",
       "3                      0.001340   0.030853        0.014149       0.002125   \n",
       "4                      0.000017   0.036614        0.005797       0.008742   \n",
       "\n",
       "   Aquimarina  Aquimarina_dash_muelleri  ...  \\\n",
       "0    0.000696                  0.000696  ...   \n",
       "1    0.000513                  0.000513  ...   \n",
       "2    0.000536                  0.000536  ...   \n",
       "3    0.001340                  0.001340  ...   \n",
       "4    0.000017                  0.000017  ...   \n",
       "\n",
       "   _leftsquarebracket_Clostridium_rightsquarebracket__dash_cocleatum  \\\n",
       "0                                           0.000696                   \n",
       "1                                           0.000513                   \n",
       "2                                           0.000536                   \n",
       "3                                           0.001340                   \n",
       "4                                           0.000017                   \n",
       "\n",
       "   _leftsquarebracket_Clostridium_rightsquarebracket__dash_innocuum_dash_group  \\\n",
       "0                                           0.000696                             \n",
       "1                                           0.000513                             \n",
       "2                                           0.000536                             \n",
       "3                                           0.001340                             \n",
       "4                                           0.000166                             \n",
       "\n",
       "   _leftsquarebracket_Eubacterium_rightsquarebracket__dash_brachy_dash_group  \\\n",
       "0                                           0.000696                           \n",
       "1                                           0.000513                           \n",
       "2                                           0.000536                           \n",
       "3                                           0.001340                           \n",
       "4                                           0.002529                           \n",
       "\n",
       "   _leftsquarebracket_Eubacterium_rightsquarebracket__dash_coprostanoligenes_dash_group  \\\n",
       "0                                           0.000696                                      \n",
       "1                                           0.000537                                      \n",
       "2                                           0.000536                                      \n",
       "3                                           0.001340                                      \n",
       "4                                           0.000017                                      \n",
       "\n",
       "   _leftsquarebracket_Eubacterium_rightsquarebracket__dash_fissicatena_dash_group  \\\n",
       "0                                           0.000696                                \n",
       "1                                           0.000513                                \n",
       "2                                           0.000536                                \n",
       "3                                           0.002369                                \n",
       "4                                           0.000017                                \n",
       "\n",
       "   _leftsquarebracket_Eubacterium_rightsquarebracket__dash_nodatum_dash_group  \\\n",
       "0                                           0.000696                            \n",
       "1                                           0.000513                            \n",
       "2                                           0.000536                            \n",
       "3                                           0.001340                            \n",
       "4                                           0.002578                            \n",
       "\n",
       "   _leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group  \\\n",
       "0                                           0.000696                                 \n",
       "1                                           0.000513                                 \n",
       "2                                           0.000536                                 \n",
       "3                                           0.001340                                 \n",
       "4                                           0.000183                                 \n",
       "\n",
       "   gut_dash_metagenome  metagenome  mouse_dash_gut_dash_metagenome  \n",
       "0             0.000696    0.000696                        0.001798  \n",
       "1             0.000513    0.000513                        0.000513  \n",
       "2             0.000536    0.000536                        0.001202  \n",
       "3             0.001340    0.001340                        0.001815  \n",
       "4             0.045397    0.000017                        0.000491  \n",
       "\n",
       "[5 rows x 116 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_16S = df_16S_meta.iloc[:,4:]\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mice_ID</th>\n",
       "      <th>Day</th>\n",
       "      <th>Group</th>\n",
       "      <th>Diet</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000000</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1000000</td>\n",
       "      <td>14.5</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Inulin</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mice_ID   Day      Group  Diet\n",
       "0  1000000   0.0  Cellulose     0\n",
       "1  1000000   0.5  Cellulose     0\n",
       "2  1000000  14.0  Cellulose     0\n",
       "3  1000000  14.5  Cellulose     0\n",
       "4        1   0.0     Inulin     1"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = df_16S_meta.iloc[:,0:4]\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the 20 most abundant taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Lactobacillus_dash_murinus</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Lactococcus</th>\n",
       "      <th>Lactobacillus</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>Alloprevotella</th>\n",
       "      <th>Muribaculaceae</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>Lachnospiraceae_dash_NK4A136_dash_group</th>\n",
       "      <th>Erysipelatoclostridium</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>Lachnoclostridium</th>\n",
       "      <th>GCA_dash_900066575</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Dorea_dash_sp_dot__dash_5_dash_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.082238</td>\n",
       "      <td>0.093941</td>\n",
       "      <td>0.130191</td>\n",
       "      <td>0.336076</td>\n",
       "      <td>0.110858</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.010021</td>\n",
       "      <td>0.036473</td>\n",
       "      <td>0.022997</td>\n",
       "      <td>0.001791</td>\n",
       "      <td>0.022697</td>\n",
       "      <td>0.001662</td>\n",
       "      <td>0.013110</td>\n",
       "      <td>0.000740</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.051865</td>\n",
       "      <td>0.019862</td>\n",
       "      <td>0.009577</td>\n",
       "      <td>0.006595</td>\n",
       "      <td>0.006772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.136595</td>\n",
       "      <td>0.052288</td>\n",
       "      <td>0.088203</td>\n",
       "      <td>0.099270</td>\n",
       "      <td>0.126145</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.022954</td>\n",
       "      <td>0.051103</td>\n",
       "      <td>0.169260</td>\n",
       "      <td>0.003593</td>\n",
       "      <td>0.041329</td>\n",
       "      <td>0.005125</td>\n",
       "      <td>0.101401</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.030730</td>\n",
       "      <td>0.005609</td>\n",
       "      <td>0.020897</td>\n",
       "      <td>0.006503</td>\n",
       "      <td>0.002247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.016578</td>\n",
       "      <td>0.094240</td>\n",
       "      <td>0.198157</td>\n",
       "      <td>0.284627</td>\n",
       "      <td>0.051968</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.003964</td>\n",
       "      <td>0.005535</td>\n",
       "      <td>0.030283</td>\n",
       "      <td>0.024240</td>\n",
       "      <td>0.025531</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.013312</td>\n",
       "      <td>0.004023</td>\n",
       "      <td>0.008377</td>\n",
       "      <td>0.001240</td>\n",
       "      <td>0.000901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.030272</td>\n",
       "      <td>0.036707</td>\n",
       "      <td>0.171329</td>\n",
       "      <td>0.409598</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.014149</td>\n",
       "      <td>0.020814</td>\n",
       "      <td>0.042389</td>\n",
       "      <td>0.010794</td>\n",
       "      <td>0.024606</td>\n",
       "      <td>0.002764</td>\n",
       "      <td>0.030853</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.015936</td>\n",
       "      <td>0.004183</td>\n",
       "      <td>0.018003</td>\n",
       "      <td>0.013521</td>\n",
       "      <td>0.002097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.100459</td>\n",
       "      <td>0.024029</td>\n",
       "      <td>0.242913</td>\n",
       "      <td>0.132299</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.005797</td>\n",
       "      <td>0.052475</td>\n",
       "      <td>0.051777</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.025668</td>\n",
       "      <td>0.031690</td>\n",
       "      <td>0.036614</td>\n",
       "      <td>0.071223</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.005889</td>\n",
       "      <td>0.013874</td>\n",
       "      <td>0.015279</td>\n",
       "      <td>0.024778</td>\n",
       "      <td>0.007378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Bacteroides_dash_acidifaciens  Lactobacillus_dash_murinus  Lachnospiraceae  \\\n",
       "0                       0.082238                    0.093941         0.130191   \n",
       "1                       0.136595                    0.052288         0.088203   \n",
       "2                       0.016578                    0.094240         0.198157   \n",
       "3                       0.030272                    0.036707         0.171329   \n",
       "4                       0.100459                    0.024029         0.242913   \n",
       "\n",
       "   Lactococcus  Lactobacillus  Bifidobacterium  Alloprevotella  \\\n",
       "0     0.336076       0.110858         0.000696        0.010021   \n",
       "1     0.099270       0.126145         0.000513        0.022954   \n",
       "2     0.284627       0.051968         0.000536        0.003964   \n",
       "3     0.409598       0.001340         0.001340        0.014149   \n",
       "4     0.132299       0.000025         0.000017        0.005797   \n",
       "\n",
       "   Muribaculaceae  Bacteroides  Roseburia  \\\n",
       "0        0.036473     0.022997   0.001791   \n",
       "1        0.051103     0.169260   0.003593   \n",
       "2        0.005535     0.030283   0.024240   \n",
       "3        0.020814     0.042389   0.010794   \n",
       "4        0.052475     0.051777   0.000524   \n",
       "\n",
       "   Lachnospiraceae_dash_NK4A136_dash_group  Erysipelatoclostridium  Alistipes  \\\n",
       "0                                 0.022697                0.001662   0.013110   \n",
       "1                                 0.041329                0.005125   0.101401   \n",
       "2                                 0.025531                0.000791   0.000536   \n",
       "3                                 0.024606                0.002764   0.030853   \n",
       "4                                 0.025668                0.031690   0.036614   \n",
       "\n",
       "   Clostridia_dash_UCG_dash_014  Akkermansia_dash_muciniphila   Blautia  \\\n",
       "0                      0.000740                      0.000696  0.051865   \n",
       "1                      0.000513                      0.000513  0.030730   \n",
       "2                      0.000536                      0.000536  0.013312   \n",
       "3                      0.001340                      0.001340  0.015936   \n",
       "4                      0.071223                      0.000017  0.005889   \n",
       "\n",
       "   Lachnoclostridium  GCA_dash_900066575  Oscillospiraceae  \\\n",
       "0           0.019862            0.009577          0.006595   \n",
       "1           0.005609            0.020897          0.006503   \n",
       "2           0.004023            0.008377          0.001240   \n",
       "3           0.004183            0.018003          0.013521   \n",
       "4           0.013874            0.015279          0.024778   \n",
       "\n",
       "   Dorea_dash_sp_dot__dash_5_dash_2  \n",
       "0                          0.006772  \n",
       "1                          0.002247  \n",
       "2                          0.000901  \n",
       "3                          0.002097  \n",
       "4                          0.007378  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topX = 20\n",
    "df_16S_T = df_16S.loc[df_meta.index].T\n",
    "df_16S_T['mean'] = df_16S_T.mean(axis=1)\n",
    "df_16S_T = df_16S_T.sort_values(by=['mean'],axis=0,ascending=False)\n",
    "df_16S_T = df_16S_T.drop('mean', axis=1)\n",
    "df_16S_topX = df_16S_T.iloc[0:topX].T\n",
    "df_16S_topX.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Bacteroides_dash_acidifaciens',\n",
       " 'Lactobacillus_dash_murinus',\n",
       " 'Lachnospiraceae',\n",
       " 'Lactococcus',\n",
       " 'Lactobacillus',\n",
       " 'Bifidobacterium',\n",
       " 'Alloprevotella',\n",
       " 'Muribaculaceae',\n",
       " 'Bacteroides',\n",
       " 'Roseburia',\n",
       " 'Lachnospiraceae_dash_NK4A136_dash_group',\n",
       " 'Erysipelatoclostridium',\n",
       " 'Alistipes',\n",
       " 'Clostridia_dash_UCG_dash_014',\n",
       " 'Akkermansia_dash_muciniphila',\n",
       " 'Blautia',\n",
       " 'Lachnoclostridium',\n",
       " 'GCA_dash_900066575',\n",
       " 'Oscillospiraceae',\n",
       " 'Dorea_dash_sp_dot__dash_5_dash_2']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_16S_topX.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize max value of 16S data to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mice_ID</th>\n",
       "      <th>Day</th>\n",
       "      <th>Group</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Lactobacillus_dash_murinus</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Lactococcus</th>\n",
       "      <th>Lactobacillus</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>...</th>\n",
       "      <th>Lachnospiraceae_dash_NK4A136_dash_group</th>\n",
       "      <th>Erysipelatoclostridium</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>Lachnoclostridium</th>\n",
       "      <th>GCA_dash_900066575</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Dorea_dash_sp_dot__dash_5_dash_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "      <td>0.090023</td>\n",
       "      <td>0.102835</td>\n",
       "      <td>0.142516</td>\n",
       "      <td>0.367892</td>\n",
       "      <td>0.121352</td>\n",
       "      <td>0.000762</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024846</td>\n",
       "      <td>0.001820</td>\n",
       "      <td>0.014351</td>\n",
       "      <td>0.000810</td>\n",
       "      <td>0.000762</td>\n",
       "      <td>0.056775</td>\n",
       "      <td>0.021742</td>\n",
       "      <td>0.010483</td>\n",
       "      <td>0.007219</td>\n",
       "      <td>0.007413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "      <td>0.149527</td>\n",
       "      <td>0.057238</td>\n",
       "      <td>0.096553</td>\n",
       "      <td>0.108668</td>\n",
       "      <td>0.138087</td>\n",
       "      <td>0.000561</td>\n",
       "      <td>...</td>\n",
       "      <td>0.045242</td>\n",
       "      <td>0.005610</td>\n",
       "      <td>0.111001</td>\n",
       "      <td>0.000561</td>\n",
       "      <td>0.000561</td>\n",
       "      <td>0.033639</td>\n",
       "      <td>0.006140</td>\n",
       "      <td>0.022875</td>\n",
       "      <td>0.007119</td>\n",
       "      <td>0.002459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000000</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "      <td>0.018148</td>\n",
       "      <td>0.103162</td>\n",
       "      <td>0.216917</td>\n",
       "      <td>0.311573</td>\n",
       "      <td>0.056888</td>\n",
       "      <td>0.000586</td>\n",
       "      <td>...</td>\n",
       "      <td>0.027948</td>\n",
       "      <td>0.000866</td>\n",
       "      <td>0.000586</td>\n",
       "      <td>0.000586</td>\n",
       "      <td>0.000586</td>\n",
       "      <td>0.014572</td>\n",
       "      <td>0.004404</td>\n",
       "      <td>0.009170</td>\n",
       "      <td>0.001357</td>\n",
       "      <td>0.000986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1000000</td>\n",
       "      <td>14.5</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "      <td>0.033137</td>\n",
       "      <td>0.040182</td>\n",
       "      <td>0.187549</td>\n",
       "      <td>0.448375</td>\n",
       "      <td>0.001467</td>\n",
       "      <td>0.001467</td>\n",
       "      <td>...</td>\n",
       "      <td>0.026936</td>\n",
       "      <td>0.003026</td>\n",
       "      <td>0.033774</td>\n",
       "      <td>0.001467</td>\n",
       "      <td>0.001467</td>\n",
       "      <td>0.017444</td>\n",
       "      <td>0.004579</td>\n",
       "      <td>0.019707</td>\n",
       "      <td>0.014801</td>\n",
       "      <td>0.002295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Inulin</td>\n",
       "      <td>1</td>\n",
       "      <td>0.109970</td>\n",
       "      <td>0.026304</td>\n",
       "      <td>0.265910</td>\n",
       "      <td>0.144823</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>...</td>\n",
       "      <td>0.028098</td>\n",
       "      <td>0.034690</td>\n",
       "      <td>0.040080</td>\n",
       "      <td>0.077966</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.006446</td>\n",
       "      <td>0.015187</td>\n",
       "      <td>0.016726</td>\n",
       "      <td>0.027124</td>\n",
       "      <td>0.008076</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mice_ID   Day      Group  Diet  Bacteroides_dash_acidifaciens  \\\n",
       "0  1000000   0.0  Cellulose     0                       0.090023   \n",
       "1  1000000   0.5  Cellulose     0                       0.149527   \n",
       "2  1000000  14.0  Cellulose     0                       0.018148   \n",
       "3  1000000  14.5  Cellulose     0                       0.033137   \n",
       "4        1   0.0     Inulin     1                       0.109970   \n",
       "\n",
       "   Lactobacillus_dash_murinus  Lachnospiraceae  Lactococcus  Lactobacillus  \\\n",
       "0                    0.102835         0.142516     0.367892       0.121352   \n",
       "1                    0.057238         0.096553     0.108668       0.138087   \n",
       "2                    0.103162         0.216917     0.311573       0.056888   \n",
       "3                    0.040182         0.187549     0.448375       0.001467   \n",
       "4                    0.026304         0.265910     0.144823       0.000027   \n",
       "\n",
       "   Bifidobacterium  ...  Lachnospiraceae_dash_NK4A136_dash_group  \\\n",
       "0         0.000762  ...                                 0.024846   \n",
       "1         0.000561  ...                                 0.045242   \n",
       "2         0.000586  ...                                 0.027948   \n",
       "3         0.001467  ...                                 0.026936   \n",
       "4         0.000018  ...                                 0.028098   \n",
       "\n",
       "   Erysipelatoclostridium  Alistipes  Clostridia_dash_UCG_dash_014  \\\n",
       "0                0.001820   0.014351                      0.000810   \n",
       "1                0.005610   0.111001                      0.000561   \n",
       "2                0.000866   0.000586                      0.000586   \n",
       "3                0.003026   0.033774                      0.001467   \n",
       "4                0.034690   0.040080                      0.077966   \n",
       "\n",
       "   Akkermansia_dash_muciniphila   Blautia  Lachnoclostridium  \\\n",
       "0                      0.000762  0.056775           0.021742   \n",
       "1                      0.000561  0.033639           0.006140   \n",
       "2                      0.000586  0.014572           0.004404   \n",
       "3                      0.001467  0.017444           0.004579   \n",
       "4                      0.000018  0.006446           0.015187   \n",
       "\n",
       "   GCA_dash_900066575  Oscillospiraceae  Dorea_dash_sp_dot__dash_5_dash_2  \n",
       "0            0.010483          0.007219                          0.007413  \n",
       "1            0.022875          0.007119                          0.002459  \n",
       "2            0.009170          0.001357                          0.000986  \n",
       "3            0.019707          0.014801                          0.002295  \n",
       "4            0.016726          0.027124                          0.008076  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterial_taxa = list(df_16S_topX.columns)\n",
    "df_16S_topX_w_meta = pd.merge(df_meta, df_16S_topX/df_16S_topX.max().max(), left_index=True, right_index=True, how='inner')\n",
    "df_16S_topX_w_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate log-derivatives (dlog(B)/dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mice_ID</th>\n",
       "      <th>Day</th>\n",
       "      <th>Group</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Lactobacillus_dash_murinus</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Lactococcus</th>\n",
       "      <th>Lactobacillus</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>...</th>\n",
       "      <th>Lachnospiraceae_dash_NK4A136_dash_group</th>\n",
       "      <th>Erysipelatoclostridium</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>Lachnoclostridium</th>\n",
       "      <th>GCA_dash_900066575</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Dorea_dash_sp_dot__dash_5_dash_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "      <td>0.129414</td>\n",
       "      <td>-0.102296</td>\n",
       "      <td>-0.101337</td>\n",
       "      <td>-0.546981</td>\n",
       "      <td>0.032627</td>\n",
       "      <td>-0.000370</td>\n",
       "      <td>...</td>\n",
       "      <td>0.043720</td>\n",
       "      <td>0.008298</td>\n",
       "      <td>0.210013</td>\n",
       "      <td>-0.000472</td>\n",
       "      <td>-0.000370</td>\n",
       "      <td>-0.049174</td>\n",
       "      <td>-0.033371</td>\n",
       "      <td>0.027355</td>\n",
       "      <td>0.000758</td>\n",
       "      <td>-0.010502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "      <td>0.108808</td>\n",
       "      <td>-0.080369</td>\n",
       "      <td>-0.082721</td>\n",
       "      <td>-0.490255</td>\n",
       "      <td>0.034232</td>\n",
       "      <td>-0.000433</td>\n",
       "      <td>...</td>\n",
       "      <td>0.037915</td>\n",
       "      <td>0.006878</td>\n",
       "      <td>0.176925</td>\n",
       "      <td>-0.000522</td>\n",
       "      <td>-0.000433</td>\n",
       "      <td>-0.043416</td>\n",
       "      <td>-0.029079</td>\n",
       "      <td>0.022269</td>\n",
       "      <td>-0.001123</td>\n",
       "      <td>-0.009323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000000</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "      <td>0.022960</td>\n",
       "      <td>-0.113892</td>\n",
       "      <td>-0.050717</td>\n",
       "      <td>0.273510</td>\n",
       "      <td>-0.104924</td>\n",
       "      <td>0.001654</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.003373</td>\n",
       "      <td>0.003735</td>\n",
       "      <td>0.054534</td>\n",
       "      <td>0.001657</td>\n",
       "      <td>0.001654</td>\n",
       "      <td>0.006743</td>\n",
       "      <td>0.001350</td>\n",
       "      <td>0.018692</td>\n",
       "      <td>0.024997</td>\n",
       "      <td>0.002756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1000000</td>\n",
       "      <td>14.5</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>0</td>\n",
       "      <td>0.037206</td>\n",
       "      <td>-0.138302</td>\n",
       "      <td>-0.066962</td>\n",
       "      <td>0.273359</td>\n",
       "      <td>-0.116838</td>\n",
       "      <td>0.001871</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000627</td>\n",
       "      <td>0.004922</td>\n",
       "      <td>0.078556</td>\n",
       "      <td>0.001868</td>\n",
       "      <td>0.001871</td>\n",
       "      <td>0.004700</td>\n",
       "      <td>-0.000689</td>\n",
       "      <td>0.023514</td>\n",
       "      <td>0.028813</td>\n",
       "      <td>0.002472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Inulin</td>\n",
       "      <td>1</td>\n",
       "      <td>0.091203</td>\n",
       "      <td>-0.017080</td>\n",
       "      <td>-0.213750</td>\n",
       "      <td>-0.143701</td>\n",
       "      <td>0.001755</td>\n",
       "      <td>-0.000178</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004772</td>\n",
       "      <td>0.130338</td>\n",
       "      <td>0.014126</td>\n",
       "      <td>-0.122993</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.035987</td>\n",
       "      <td>-0.016806</td>\n",
       "      <td>0.014997</td>\n",
       "      <td>-0.036714</td>\n",
       "      <td>-0.003546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mice_ID   Day      Group  Diet  Bacteroides_dash_acidifaciens  \\\n",
       "0  1000000   0.0  Cellulose     0                       0.129414   \n",
       "1  1000000   0.5  Cellulose     0                       0.108808   \n",
       "2  1000000  14.0  Cellulose     0                       0.022960   \n",
       "3  1000000  14.5  Cellulose     0                       0.037206   \n",
       "4        1   0.0     Inulin     1                       0.091203   \n",
       "\n",
       "   Lactobacillus_dash_murinus  Lachnospiraceae  Lactococcus  Lactobacillus  \\\n",
       "0                   -0.102296        -0.101337    -0.546981       0.032627   \n",
       "1                   -0.080369        -0.082721    -0.490255       0.034232   \n",
       "2                   -0.113892        -0.050717     0.273510      -0.104924   \n",
       "3                   -0.138302        -0.066962     0.273359      -0.116838   \n",
       "4                   -0.017080        -0.213750    -0.143701       0.001755   \n",
       "\n",
       "   Bifidobacterium  ...  Lachnospiraceae_dash_NK4A136_dash_group  \\\n",
       "0        -0.000370  ...                                 0.043720   \n",
       "1        -0.000433  ...                                 0.037915   \n",
       "2         0.001654  ...                                -0.003373   \n",
       "3         0.001871  ...                                -0.000627   \n",
       "4        -0.000178  ...                                -0.004772   \n",
       "\n",
       "   Erysipelatoclostridium  Alistipes  Clostridia_dash_UCG_dash_014  \\\n",
       "0                0.008298   0.210013                     -0.000472   \n",
       "1                0.006878   0.176925                     -0.000522   \n",
       "2                0.003735   0.054534                      0.001657   \n",
       "3                0.004922   0.078556                      0.001868   \n",
       "4                0.130338   0.014126                     -0.122993   \n",
       "\n",
       "   Akkermansia_dash_muciniphila   Blautia  Lachnoclostridium  \\\n",
       "0                     -0.000370 -0.049174          -0.033371   \n",
       "1                     -0.000433 -0.043416          -0.029079   \n",
       "2                      0.001654  0.006743           0.001350   \n",
       "3                      0.001871  0.004700          -0.000689   \n",
       "4                      0.000176  0.035987          -0.016806   \n",
       "\n",
       "   GCA_dash_900066575  Oscillospiraceae  Dorea_dash_sp_dot__dash_5_dash_2  \n",
       "0            0.027355          0.000758                         -0.010502  \n",
       "1            0.022269         -0.001123                         -0.009323  \n",
       "2            0.018692          0.024997                          0.002756  \n",
       "3            0.023514          0.028813                          0.002472  \n",
       "4            0.014997         -0.036714                         -0.003546  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_deriv = deepcopy(df_16S_topX_w_meta)\n",
    "for curr_sub in set(df_deriv.Mice_ID):\n",
    "    curr_df = df_deriv[df_deriv.Mice_ID==curr_sub].sort_values(by='Day')\n",
    "    if len(curr_df.index) == 0:\n",
    "        continue\n",
    "    for taxon in bacterial_taxa:\n",
    "        xdata = np.array(curr_df['Day'])\n",
    "        ydata = np.array(curr_df[taxon])\n",
    "        cs = CubicSpline(xdata, ydata)\n",
    "        csd1 = cs.derivative(nu=1)\n",
    "        ydata_d1 = csd1(xdata)\n",
    "        df_deriv.loc[df_deriv.Mice_ID==curr_sub, taxon] = ydata_d1\n",
    "df_deriv.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct regression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 1.25848493,  1.05551206,  0.20986347, ..., -0.02867623,\n",
       "       -0.00316822, -0.0143775 ])"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ymat should be samples by bacteria\n",
    "Ymat = df_deriv[bacterial_taxa].values\n",
    "Ymat = Ymat.flatten(order='F')\n",
    "Ymat = StandardScaler().fit_transform(Ymat.reshape(-1,1)).reshape(1,-1)[0] # standardize\n",
    "Ymat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1700"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Ymat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.00000000e+00, 0.00000000e+00, 9.00231553e-02, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 0.00000000e+00, 1.49526782e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 0.00000000e+00, 1.81479355e-02, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       ...,\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        1.12075336e-03, 4.03471209e-03, 2.03773338e-04],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        1.73072191e-03, 8.76599410e-04, 1.05641467e-03],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        3.38165644e-03, 7.32692229e-04, 5.86153784e-03]])"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Xmat = np.zeros(shape=(topX*len(df_deriv.index), (topX+2)*topX))\n",
    "for k in np.arange(topX):\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)] = 1\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+1] = df_deriv.Diet.values\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+2:(k+1)*(topX+2)] = df_16S_topX_w_meta[bacterial_taxa].values\n",
    "Xmat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write stan program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "json_str = '{\\n\"N\" : %d,\\n'%(len(Ymat))\n",
    "\n",
    "# y variable\n",
    "json_str += '\\\"dlogX\\\" : [%s],\\n'%(','.join(list(Ymat.astype(str))))\n",
    "\n",
    "# x variable\n",
    "for k1,c1 in enumerate(bacterial_taxa):\n",
    "    # growth rate\n",
    "    json_str += '\\\"growth_rate_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)].astype(str))))\n",
    "    # diet response\n",
    "    json_str += '\\\"inulin_response_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)+1].astype(str))))\n",
    "    # bacterial interactions\n",
    "    for k2,c2 in enumerate(bacterial_taxa):\n",
    "        v = list(Xmat[:,k1*(topX+2)+2+k2].astype(str))\n",
    "        json_str += '\\\"pairwise_interaction_%s_%s\\\" : [%s]'%(c1,c2,','.join(v))\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            json_str += '\\n}'\n",
    "        else:\n",
    "            json_str += ',\\n'\n",
    "#print(json_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.data.json\", \"w\")\n",
    "text_file.write(\"%s\" % json_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# data block\n",
    "model_str = 'data {\\n'\n",
    "model_str += '\\tint<lower=0> N;\\n'\n",
    "model_str += '\\tvector[N] dlogX;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\tvector[N] growth_rate_%s;\\n'%(c1)\n",
    "    model_str += '\\tvector[N] inulin_response_%s;\\n'%(c1)\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tvector[N] pairwise_interaction_%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'\n",
    "\n",
    "# parameter block\n",
    "model_str += 'parameters {\\n\\treal<lower=0,upper=1> sigma;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\treal alpha__%s;\\n'%(c1) # growth rate\n",
    "    model_str += '\\treal epsilon__%s;\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\treal beta__%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'       \n",
    "        \n",
    "# model block\n",
    "model_str += 'model {\\n\\tsigma ~ uniform(0,1);\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\talpha__%s ~ normal(0,1);\\n'%(c1) # growth rate\n",
    "    model_str += '\\tepsilon__%s ~ normal(0,1);\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tbeta__%s_%s ~ normal(0,1);\\n'%(c1,c2)\n",
    "model_str += '\\tdlogX ~ normal('\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += 'alpha__%s*growth_rate_%s+'%(c1,c1) # growth rate\n",
    "    model_str += 'epsilon__%s*inulin_response_%s+'%(c1,c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s'%(c1,c2,c1,c2)\n",
    "        else:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s+'%(c1,c2,c1,c2)\n",
    "model_str += ', sigma);\\n}'\n",
    "#print(model_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.stan\", \"w\")\n",
    "text_file.write(\"%s\" % model_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot stan output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hpd_grid(sample, alpha=0.05, roundto=2):\n",
    "    \"\"\"Calculate highest posterior density (HPD) of array for given alpha. \n",
    "    The HPD is the minimum width Bayesian credible interval (BCI). \n",
    "    The function works for multimodal distributions, returning more than one mode\n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    sample : Numpy array or python list\n",
    "        An array containing MCMC samples\n",
    "    alpha : float\n",
    "        Desired probability of type I error (defaults to 0.05)\n",
    "    roundto: integer\n",
    "        Number of digits after the decimal point for the results\n",
    "    Returns\n",
    "    ----------\n",
    "    hpd: array with the lower \n",
    "          \n",
    "    \"\"\"\n",
    "    sample = np.asarray(sample)\n",
    "    sample = sample[~np.isnan(sample)]\n",
    "    # get upper and lower bounds\n",
    "    l = np.min(sample)\n",
    "    u = np.max(sample)\n",
    "    density = kde.gaussian_kde(sample)\n",
    "    x = np.linspace(l, u, 2000)\n",
    "    y = density.evaluate(x)\n",
    "    #y = density.evaluate(x, l, u) waitting for PR to be accepted\n",
    "    xy_zipped = zip(x, y/np.sum(y))\n",
    "    xy = sorted(xy_zipped, key=lambda x: x[1], reverse=True)\n",
    "    xy_cum_sum = 0\n",
    "    hdv = []\n",
    "    for val in xy:\n",
    "        xy_cum_sum += val[1]\n",
    "        hdv.append(val[0])\n",
    "        if xy_cum_sum >= (1-alpha):\n",
    "            break\n",
    "    hdv.sort()\n",
    "    diff = (u-l)/20  # differences of 5%\n",
    "    hpd = []\n",
    "    hpd.append(round(min(hdv), roundto))\n",
    "    for i in range(1, len(hdv)):\n",
    "        if hdv[i]-hdv[i-1] >= diff:\n",
    "            hpd.append(round(hdv[i-1], roundto))\n",
    "            hpd.append(round(hdv[i], roundto))\n",
    "    hpd.append(round(max(hdv), roundto))\n",
    "    ite = iter(hpd)\n",
    "    hpd = list(zip(ite, ite))\n",
    "    modes = []\n",
    "    for value in hpd:\n",
    "        x_hpd = x[(x > value[0]) & (x < value[1])]\n",
    "        y_hpd = y[(x > value[0]) & (x < value[1])]\n",
    "        modes.append(round(x_hpd[np.argmax(y_hpd)], roundto))\n",
    "    return hpd, x, y, modes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = az.from_cmdstan([\"output_%d.csv\"%(i) for i in np.arange(1,4)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Muribaculaceae</td>\n",
       "      <td>-0.16</td>\n",
       "      <td>0.15</td>\n",
       "      <td>-0.005</td>\n",
       "      <td>-0.044859</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bacteroides_dash_acidifaciens</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.145</td>\n",
       "      <td>1.835619</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Akkermansia_dash_muciniphila</td>\n",
       "      <td>-0.27</td>\n",
       "      <td>0.04</td>\n",
       "      <td>-0.115</td>\n",
       "      <td>-1.416896</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Faecalibaculum</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-0.040</td>\n",
       "      <td>-0.531562</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Parasutterella</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-0.040</td>\n",
       "      <td>-0.540205</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Type                           Taxa  Left  Right  Middle  \\\n",
       "0  basal_growth_rate                 Muribaculaceae -0.16   0.15  -0.005   \n",
       "1  basal_growth_rate  Bacteroides_dash_acidifaciens -0.01   0.30   0.145   \n",
       "2  basal_growth_rate   Akkermansia_dash_muciniphila -0.27   0.04  -0.115   \n",
       "3  basal_growth_rate                 Faecalibaculum -0.20   0.12  -0.040   \n",
       "4  basal_growth_rate                 Parasutterella -0.20   0.12  -0.040   \n",
       "\n",
       "        SNR  Significant  \n",
       "0 -0.044859        False  \n",
       "1  1.835619        False  \n",
       "2 -1.416896        False  \n",
       "3 -0.531562        False  \n",
       "4 -0.540205        False  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines = []\n",
    "    \n",
    "# basal growth rate\n",
    "for c in bacterial_taxa:\n",
    "    var = 'alpha__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['basal_growth_rate', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# inulin response\n",
    "for c in bacterial_taxa:\n",
    "    var = 'epsilon__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['inulin_response', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "\n",
    "# pairwise interactions\n",
    "for c1 in bacterial_taxa:\n",
    "    for c2 in bacterial_taxa:\n",
    "        var = 'beta__%s_%s'%(c1,c2)\n",
    "        data = []\n",
    "        for i in np.arange(0,3):\n",
    "            data.extend(list(fit.posterior[var][i].values))\n",
    "        hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "        assert len(hpd_mu) == 1\n",
    "        (x0, x1) = hpd_mu[0]\n",
    "        lines.append(['pairwise_interaction', (c1,c2), x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "df_stan_output_summary = pd.DataFrame(lines, columns = ['Type','Taxa','Left','Right','Middle','SNR','Significant'])\n",
    "df_stan_output_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stan_output_summary.to_excel('bayesian_regression_summary_top20.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>-0.99</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>-0.505</td>\n",
       "      <td>-2.050357</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Bifidobacterium</td>\n",
       "      <td>1.57</td>\n",
       "      <td>2.19</td>\n",
       "      <td>1.880</td>\n",
       "      <td>12.046869</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Anaerostipes</td>\n",
       "      <td>0.94</td>\n",
       "      <td>1.56</td>\n",
       "      <td>1.250</td>\n",
       "      <td>7.995026</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.425</td>\n",
       "      <td>2.720541</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>-0.66</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>-0.350</td>\n",
       "      <td>-2.239965</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Coprococcus</td>\n",
       "      <td>-0.65</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>-0.340</td>\n",
       "      <td>-2.163012</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Blautia', 'Blautia')</td>\n",
       "      <td>-3.91</td>\n",
       "      <td>-1.19</td>\n",
       "      <td>-2.550</td>\n",
       "      <td>-3.696627</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Blautia', 'Faecalibacterium')</td>\n",
       "      <td>-3.30</td>\n",
       "      <td>-0.11</td>\n",
       "      <td>-1.705</td>\n",
       "      <td>-2.138517</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Blautia', 'Bifidobacterium')</td>\n",
       "      <td>-3.51</td>\n",
       "      <td>-0.87</td>\n",
       "      <td>-2.190</td>\n",
       "      <td>-3.272662</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bifidobacterium', 'Bifidobacterium')</td>\n",
       "      <td>1.35</td>\n",
       "      <td>3.97</td>\n",
       "      <td>2.660</td>\n",
       "      <td>3.995891</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Prevotella', 'Faecalibacterium')</td>\n",
       "      <td>0.23</td>\n",
       "      <td>3.40</td>\n",
       "      <td>1.815</td>\n",
       "      <td>2.260957</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Prevotella', 'Prevotella')</td>\n",
       "      <td>-6.48</td>\n",
       "      <td>-4.34</td>\n",
       "      <td>-5.410</td>\n",
       "      <td>-9.988830</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Type                                    Taxa  Left  \\\n",
       "6       basal_growth_rate                              Prevotella -0.99   \n",
       "22        inulin_response                         Bifidobacterium  1.57   \n",
       "24        inulin_response                            Anaerostipes  0.94   \n",
       "26        inulin_response                              Prevotella  0.12   \n",
       "28        inulin_response                             Bacteroides -0.66   \n",
       "33        inulin_response                             Coprococcus -0.65   \n",
       "40   pairwise_interaction                  ('Blautia', 'Blautia') -3.91   \n",
       "41   pairwise_interaction         ('Blautia', 'Faecalibacterium') -3.30   \n",
       "42   pairwise_interaction          ('Blautia', 'Bifidobacterium') -3.51   \n",
       "82   pairwise_interaction  ('Bifidobacterium', 'Bifidobacterium')  1.35   \n",
       "161  pairwise_interaction      ('Prevotella', 'Faecalibacterium')  0.23   \n",
       "166  pairwise_interaction            ('Prevotella', 'Prevotella') -6.48   \n",
       "\n",
       "     Right  Middle        SNR  Significant  \n",
       "6    -0.02  -0.505  -2.050357         True  \n",
       "22    2.19   1.880  12.046869         True  \n",
       "24    1.56   1.250   7.995026         True  \n",
       "26    0.73   0.425   2.720541         True  \n",
       "28   -0.04  -0.350  -2.239965         True  \n",
       "33   -0.03  -0.340  -2.163012         True  \n",
       "40   -1.19  -2.550  -3.696627         True  \n",
       "41   -0.11  -1.705  -2.138517         True  \n",
       "42   -0.87  -2.190  -3.272662         True  \n",
       "82    3.97   2.660   3.995891         True  \n",
       "161   3.40   1.815   2.260957         True  \n",
       "166  -4.34  -5.410  -9.988830         True  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stan_output_summary = pd.read_excel('bayesian_regression_summary_top20.xlsx', index_col=0)\n",
    "df_stan_output_summary = df_stan_output_summary[df_stan_output_summary.Significant==True]\n",
    "df_stan_output_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
