{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from copy import deepcopy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from scipy.interpolate import CubicSpline\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import warnings\n",
    "import itertools\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "import statsmodels.stats.api as sms\n",
    "import scipy.stats.kde as kde"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read 16S data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Abiotrophia_defectiva</th>\n",
       "      <th>Abiotrophia_sp_dot__HMSC24B09</th>\n",
       "      <th>Absiella_dolichum</th>\n",
       "      <th>Acetanaerobacterium_elongatum</th>\n",
       "      <th>Acetitomaculum_ruminis</th>\n",
       "      <th>Acetivibrio_cellulolyticus</th>\n",
       "      <th>Acetivibrio_ethanolgignens</th>\n",
       "      <th>Acetoanaerobium_sticklandii</th>\n",
       "      <th>Acetobacter_nitrogenifigens</th>\n",
       "      <th>Acetobacter_papayae</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__combesii</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__eligens</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__hallii</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__siraeum</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__yurii</th>\n",
       "      <th>_leftsquarebracket_Haemophilus_rightsquarebracket__ducreyi</th>\n",
       "      <th>_leftsquarebracket_Hallella_rightsquarebracket__seregens</th>\n",
       "      <th>_leftsquarebracket_Propionibacterium_rightsquarebracket__namnetense</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__gnavus</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__torques</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G212_SS001685</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000097</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.013440</td>\n",
       "      <td>0.012065</td>\n",
       "      <td>0.000423</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001914</td>\n",
       "      <td>0.005027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001781</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000201</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004616</td>\n",
       "      <td>0.020981</td>\n",
       "      <td>0.000221</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002235</td>\n",
       "      <td>0.010386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001782</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000522</td>\n",
       "      <td>0.037412</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003169</td>\n",
       "      <td>0.008648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001783</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001219</td>\n",
       "      <td>0.047010</td>\n",
       "      <td>0.001616</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003124</td>\n",
       "      <td>0.012215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001784</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000611</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000855</td>\n",
       "      <td>0.023347</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002769</td>\n",
       "      <td>0.019837</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3739 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Abiotrophia_defectiva  Abiotrophia_sp_dot__HMSC24B09  \\\n",
       "SampleID                                                              \n",
       "G212_SS001685                    0.0                       0.000009   \n",
       "G212_SS001781                    0.0                       0.000000   \n",
       "G212_SS001782                    0.0                       0.000003   \n",
       "G212_SS001783                    0.0                       0.000000   \n",
       "G212_SS001784                    0.0                       0.000006   \n",
       "\n",
       "               Absiella_dolichum  Acetanaerobacterium_elongatum  \\\n",
       "SampleID                                                          \n",
       "G212_SS001685           0.000018                            0.0   \n",
       "G212_SS001781           0.000201                            0.0   \n",
       "G212_SS001782           0.000018                            0.0   \n",
       "G212_SS001783           0.000027                            0.0   \n",
       "G212_SS001784           0.000611                            0.0   \n",
       "\n",
       "               Acetitomaculum_ruminis  Acetivibrio_cellulolyticus  \\\n",
       "SampleID                                                            \n",
       "G212_SS001685                0.000000                         0.0   \n",
       "G212_SS001781                0.000000                         0.0   \n",
       "G212_SS001782                0.000000                         0.0   \n",
       "G212_SS001783                0.000000                         0.0   \n",
       "G212_SS001784                0.000003                         0.0   \n",
       "\n",
       "               Acetivibrio_ethanolgignens  Acetoanaerobium_sticklandii  \\\n",
       "SampleID                                                                 \n",
       "G212_SS001685                    0.000097                          0.0   \n",
       "G212_SS001781                    0.000046                          0.0   \n",
       "G212_SS001782                    0.000036                          0.0   \n",
       "G212_SS001783                    0.000042                          0.0   \n",
       "G212_SS001784                    0.000047                          0.0   \n",
       "\n",
       "               Acetobacter_nitrogenifigens  Acetobacter_papayae  ...  \\\n",
       "SampleID                                                         ...   \n",
       "G212_SS001685                          0.0                  0.0  ...   \n",
       "G212_SS001781                          0.0                  0.0  ...   \n",
       "G212_SS001782                          0.0                  0.0  ...   \n",
       "G212_SS001783                          0.0                  0.0  ...   \n",
       "G212_SS001784                          0.0                  0.0  ...   \n",
       "\n",
       "               _leftsquarebracket_Eubacterium_rightsquarebracket__combesii  \\\n",
       "SampleID                                                                     \n",
       "G212_SS001685                                                0.0             \n",
       "G212_SS001781                                                0.0             \n",
       "G212_SS001782                                                0.0             \n",
       "G212_SS001783                                                0.0             \n",
       "G212_SS001784                                                0.0             \n",
       "\n",
       "               _leftsquarebracket_Eubacterium_rightsquarebracket__eligens  \\\n",
       "SampleID                                                                    \n",
       "G212_SS001685                                           0.013440            \n",
       "G212_SS001781                                           0.004616            \n",
       "G212_SS001782                                           0.000522            \n",
       "G212_SS001783                                           0.001219            \n",
       "G212_SS001784                                           0.000855            \n",
       "\n",
       "               _leftsquarebracket_Eubacterium_rightsquarebracket__hallii  \\\n",
       "SampleID                                                                   \n",
       "G212_SS001685                                           0.012065           \n",
       "G212_SS001781                                           0.020981           \n",
       "G212_SS001782                                           0.037412           \n",
       "G212_SS001783                                           0.047010           \n",
       "G212_SS001784                                           0.023347           \n",
       "\n",
       "               _leftsquarebracket_Eubacterium_rightsquarebracket__siraeum  \\\n",
       "SampleID                                                                    \n",
       "G212_SS001685                                           0.000423            \n",
       "G212_SS001781                                           0.000221            \n",
       "G212_SS001782                                           0.000629            \n",
       "G212_SS001783                                           0.001616            \n",
       "G212_SS001784                                           0.000116            \n",
       "\n",
       "               _leftsquarebracket_Eubacterium_rightsquarebracket__yurii  \\\n",
       "SampleID                                                                  \n",
       "G212_SS001685                                                0.0          \n",
       "G212_SS001781                                                0.0          \n",
       "G212_SS001782                                                0.0          \n",
       "G212_SS001783                                                0.0          \n",
       "G212_SS001784                                                0.0          \n",
       "\n",
       "               _leftsquarebracket_Haemophilus_rightsquarebracket__ducreyi  \\\n",
       "SampleID                                                                    \n",
       "G212_SS001685                                           0.000000            \n",
       "G212_SS001781                                           0.000016            \n",
       "G212_SS001782                                           0.000000            \n",
       "G212_SS001783                                           0.000000            \n",
       "G212_SS001784                                           0.000000            \n",
       "\n",
       "               _leftsquarebracket_Hallella_rightsquarebracket__seregens  \\\n",
       "SampleID                                                                  \n",
       "G212_SS001685                                           0.000000          \n",
       "G212_SS001781                                           0.000000          \n",
       "G212_SS001782                                           0.000000          \n",
       "G212_SS001783                                           0.000000          \n",
       "G212_SS001784                                           0.000006          \n",
       "\n",
       "               _leftsquarebracket_Propionibacterium_rightsquarebracket__namnetense  \\\n",
       "SampleID                                                                             \n",
       "G212_SS001685                                                0.0                     \n",
       "G212_SS001781                                                0.0                     \n",
       "G212_SS001782                                                0.0                     \n",
       "G212_SS001783                                                0.0                     \n",
       "G212_SS001784                                                0.0                     \n",
       "\n",
       "               _leftsquarebracket_Ruminococcus_rightsquarebracket__gnavus  \\\n",
       "SampleID                                                                    \n",
       "G212_SS001685                                           0.001914            \n",
       "G212_SS001781                                           0.002235            \n",
       "G212_SS001782                                           0.003169            \n",
       "G212_SS001783                                           0.003124            \n",
       "G212_SS001784                                           0.002769            \n",
       "\n",
       "               _leftsquarebracket_Ruminococcus_rightsquarebracket__torques  \n",
       "SampleID                                                                    \n",
       "G212_SS001685                                           0.005027            \n",
       "G212_SS001781                                           0.010386            \n",
       "G212_SS001782                                           0.008648            \n",
       "G212_SS001783                                           0.012215            \n",
       "G212_SS001784                                           0.019837            \n",
       "\n",
       "[5 rows x 3739 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_16S = pd.read_excel('./16S_relative_abundance.xlsx', index_col=0)\n",
    "df_16S.columns = [c.replace('/','_slash_').replace(' ','_space_').replace('[','_leftsquarebracket_').replace(']','_rightsquarebracket_').replace('-','_dash_').replace('.','_dot_').replace('(','_leftroundbracket').replace(')','_rightroundbracket_') for c in df_16S.columns]\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add pseudocount for zeros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Abiotrophia_defectiva</th>\n",
       "      <th>Abiotrophia_sp_dot__HMSC24B09</th>\n",
       "      <th>Absiella_dolichum</th>\n",
       "      <th>Acetanaerobacterium_elongatum</th>\n",
       "      <th>Acetitomaculum_ruminis</th>\n",
       "      <th>Acetivibrio_cellulolyticus</th>\n",
       "      <th>Acetivibrio_ethanolgignens</th>\n",
       "      <th>Acetoanaerobium_sticklandii</th>\n",
       "      <th>Acetobacter_nitrogenifigens</th>\n",
       "      <th>Acetobacter_papayae</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__combesii</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__eligens</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__hallii</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__siraeum</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__yurii</th>\n",
       "      <th>_leftsquarebracket_Haemophilus_rightsquarebracket__ducreyi</th>\n",
       "      <th>_leftsquarebracket_Hallella_rightsquarebracket__seregens</th>\n",
       "      <th>_leftsquarebracket_Propionibacterium_rightsquarebracket__namnetense</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__gnavus</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__torques</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G212_SS001685</th>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000097</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.013440</td>\n",
       "      <td>0.012065</td>\n",
       "      <td>0.000423</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.001914</td>\n",
       "      <td>0.005027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001781</th>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000201</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.004616</td>\n",
       "      <td>0.020981</td>\n",
       "      <td>0.000221</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.002235</td>\n",
       "      <td>0.010386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001782</th>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000522</td>\n",
       "      <td>0.037412</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.003169</td>\n",
       "      <td>0.008648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001783</th>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.001219</td>\n",
       "      <td>0.047010</td>\n",
       "      <td>0.001616</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.003124</td>\n",
       "      <td>0.012215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001784</th>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000611</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000855</td>\n",
       "      <td>0.023347</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.002769</td>\n",
       "      <td>0.019837</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3739 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Abiotrophia_defectiva  Abiotrophia_sp_dot__HMSC24B09  \\\n",
       "SampleID                                                              \n",
       "G212_SS001685               0.000009                       0.000009   \n",
       "G212_SS001781               0.000002                       0.000002   \n",
       "G212_SS001782               0.000003                       0.000003   \n",
       "G212_SS001783               0.000004                       0.000004   \n",
       "G212_SS001784               0.000003                       0.000006   \n",
       "\n",
       "               Absiella_dolichum  Acetanaerobacterium_elongatum  \\\n",
       "SampleID                                                          \n",
       "G212_SS001685           0.000018                       0.000009   \n",
       "G212_SS001781           0.000201                       0.000002   \n",
       "G212_SS001782           0.000018                       0.000003   \n",
       "G212_SS001783           0.000027                       0.000004   \n",
       "G212_SS001784           0.000611                       0.000003   \n",
       "\n",
       "               Acetitomaculum_ruminis  Acetivibrio_cellulolyticus  \\\n",
       "SampleID                                                            \n",
       "G212_SS001685                0.000009                    0.000009   \n",
       "G212_SS001781                0.000002                    0.000002   \n",
       "G212_SS001782                0.000003                    0.000003   \n",
       "G212_SS001783                0.000004                    0.000004   \n",
       "G212_SS001784                0.000003                    0.000003   \n",
       "\n",
       "               Acetivibrio_ethanolgignens  Acetoanaerobium_sticklandii  \\\n",
       "SampleID                                                                 \n",
       "G212_SS001685                    0.000097                     0.000009   \n",
       "G212_SS001781                    0.000046                     0.000002   \n",
       "G212_SS001782                    0.000036                     0.000003   \n",
       "G212_SS001783                    0.000042                     0.000004   \n",
       "G212_SS001784                    0.000047                     0.000003   \n",
       "\n",
       "               Acetobacter_nitrogenifigens  Acetobacter_papayae  ...  \\\n",
       "SampleID                                                         ...   \n",
       "G212_SS001685                     0.000009             0.000009  ...   \n",
       "G212_SS001781                     0.000002             0.000002  ...   \n",
       "G212_SS001782                     0.000003             0.000003  ...   \n",
       "G212_SS001783                     0.000004             0.000004  ...   \n",
       "G212_SS001784                     0.000003             0.000003  ...   \n",
       "\n",
       "               _leftsquarebracket_Eubacterium_rightsquarebracket__combesii  \\\n",
       "SampleID                                                                     \n",
       "G212_SS001685                                           0.000009             \n",
       "G212_SS001781                                           0.000002             \n",
       "G212_SS001782                                           0.000003             \n",
       "G212_SS001783                                           0.000004             \n",
       "G212_SS001784                                           0.000003             \n",
       "\n",
       "               _leftsquarebracket_Eubacterium_rightsquarebracket__eligens  \\\n",
       "SampleID                                                                    \n",
       "G212_SS001685                                           0.013440            \n",
       "G212_SS001781                                           0.004616            \n",
       "G212_SS001782                                           0.000522            \n",
       "G212_SS001783                                           0.001219            \n",
       "G212_SS001784                                           0.000855            \n",
       "\n",
       "               _leftsquarebracket_Eubacterium_rightsquarebracket__hallii  \\\n",
       "SampleID                                                                   \n",
       "G212_SS001685                                           0.012065           \n",
       "G212_SS001781                                           0.020981           \n",
       "G212_SS001782                                           0.037412           \n",
       "G212_SS001783                                           0.047010           \n",
       "G212_SS001784                                           0.023347           \n",
       "\n",
       "               _leftsquarebracket_Eubacterium_rightsquarebracket__siraeum  \\\n",
       "SampleID                                                                    \n",
       "G212_SS001685                                           0.000423            \n",
       "G212_SS001781                                           0.000221            \n",
       "G212_SS001782                                           0.000629            \n",
       "G212_SS001783                                           0.001616            \n",
       "G212_SS001784                                           0.000116            \n",
       "\n",
       "               _leftsquarebracket_Eubacterium_rightsquarebracket__yurii  \\\n",
       "SampleID                                                                  \n",
       "G212_SS001685                                           0.000009          \n",
       "G212_SS001781                                           0.000002          \n",
       "G212_SS001782                                           0.000003          \n",
       "G212_SS001783                                           0.000004          \n",
       "G212_SS001784                                           0.000003          \n",
       "\n",
       "               _leftsquarebracket_Haemophilus_rightsquarebracket__ducreyi  \\\n",
       "SampleID                                                                    \n",
       "G212_SS001685                                           0.000009            \n",
       "G212_SS001781                                           0.000016            \n",
       "G212_SS001782                                           0.000003            \n",
       "G212_SS001783                                           0.000004            \n",
       "G212_SS001784                                           0.000003            \n",
       "\n",
       "               _leftsquarebracket_Hallella_rightsquarebracket__seregens  \\\n",
       "SampleID                                                                  \n",
       "G212_SS001685                                           0.000009          \n",
       "G212_SS001781                                           0.000002          \n",
       "G212_SS001782                                           0.000003          \n",
       "G212_SS001783                                           0.000004          \n",
       "G212_SS001784                                           0.000006          \n",
       "\n",
       "               _leftsquarebracket_Propionibacterium_rightsquarebracket__namnetense  \\\n",
       "SampleID                                                                             \n",
       "G212_SS001685                                           0.000009                     \n",
       "G212_SS001781                                           0.000002                     \n",
       "G212_SS001782                                           0.000003                     \n",
       "G212_SS001783                                           0.000004                     \n",
       "G212_SS001784                                           0.000003                     \n",
       "\n",
       "               _leftsquarebracket_Ruminococcus_rightsquarebracket__gnavus  \\\n",
       "SampleID                                                                    \n",
       "G212_SS001685                                           0.001914            \n",
       "G212_SS001781                                           0.002235            \n",
       "G212_SS001782                                           0.003169            \n",
       "G212_SS001783                                           0.003124            \n",
       "G212_SS001784                                           0.002769            \n",
       "\n",
       "               _leftsquarebracket_Ruminococcus_rightsquarebracket__torques  \n",
       "SampleID                                                                    \n",
       "G212_SS001685                                           0.005027            \n",
       "G212_SS001781                                           0.010386            \n",
       "G212_SS001782                                           0.008648            \n",
       "G212_SS001783                                           0.012215            \n",
       "G212_SS001784                                           0.019837            \n",
       "\n",
       "[5 rows x 3739 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for sample_id in df_16S.index:\n",
    "    sample = np.array(df_16S.loc[sample_id])\n",
    "    minval = np.min(sample[np.nonzero(sample)]) # minimum non-zero value\n",
    "    sample[sample==0] = minval\n",
    "    df_16S.loc[sample_id] = sample\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G212_SS001685</th>\n",
       "      <td>101-030</td>\n",
       "      <td>50.135417</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001781</th>\n",
       "      <td>101-007</td>\n",
       "      <td>50.052778</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001782</th>\n",
       "      <td>101-028</td>\n",
       "      <td>41.486111</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001783</th>\n",
       "      <td>101-028</td>\n",
       "      <td>44.583333</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001784</th>\n",
       "      <td>101-039</td>\n",
       "      <td>22.177083</td>\n",
       "      <td>2.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Subject        Day  Diet\n",
       "Sample Name                            \n",
       "G212_SS001685  101-030  50.135417   0.0\n",
       "G212_SS001781  101-007  50.052778   0.0\n",
       "G212_SS001782  101-028  41.486111   0.0\n",
       "G212_SS001783  101-028  44.583333   0.0\n",
       "G212_SS001784  101-039  22.177083   2.5"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_excel('fos_metadata.xlsx', sheet_name='metadata').set_index('Sample Name')\n",
    "df_meta = df_meta.rename({'study_day':'Day', 'dose_g':'Diet'}, axis=1)\n",
    "df_meta = df_meta[['Subject','Day','Diet']]\n",
    "df_meta = df_meta[df_meta.Subject.notnull()]\n",
    "\n",
    "# remove subject that has single data\n",
    "subject_to_remove = []\n",
    "for sub in set(df_meta.Subject):\n",
    "    if len(df_meta[df_meta.Subject==sub])==1:\n",
    "        subject_to_remove.append(sub)\n",
    "df_meta = df_meta[~df_meta.Subject.isin(subject_to_remove)]\n",
    "\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the 20 most abundant taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Faecalibacterium_prausnitzii</th>\n",
       "      <th>Dorea_longicatena</th>\n",
       "      <th>Fusicatenibacter_saccharivorans</th>\n",
       "      <th>Anaerostipes_hadrus</th>\n",
       "      <th>Bifidobacterium_adolescentis</th>\n",
       "      <th>Prevotella_copri</th>\n",
       "      <th>Blautia_obeum</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__hallii</th>\n",
       "      <th>Roseburia_faecis</th>\n",
       "      <th>Akkermansia_muciniphila</th>\n",
       "      <th>Bifidobacterium_longum</th>\n",
       "      <th>Coprococcus_comes</th>\n",
       "      <th>Blautia_wexlerae</th>\n",
       "      <th>Collinsella_aerofaciens</th>\n",
       "      <th>Gemmiger_formicilis</th>\n",
       "      <th>Ruminococcus_bicirculans</th>\n",
       "      <th>Bifidobacterium_bifidum</th>\n",
       "      <th>Roseburia_intestinalis</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__torques</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G212_SS001685</th>\n",
       "      <td>0.109473</td>\n",
       "      <td>0.244980</td>\n",
       "      <td>0.044713</td>\n",
       "      <td>0.023732</td>\n",
       "      <td>0.025135</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.050613</td>\n",
       "      <td>0.013379</td>\n",
       "      <td>0.012065</td>\n",
       "      <td>0.109579</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.007787</td>\n",
       "      <td>0.008149</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.036900</td>\n",
       "      <td>0.008722</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.008828</td>\n",
       "      <td>0.005027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001781</th>\n",
       "      <td>0.124860</td>\n",
       "      <td>0.061379</td>\n",
       "      <td>0.094882</td>\n",
       "      <td>0.086011</td>\n",
       "      <td>0.033223</td>\n",
       "      <td>0.113808</td>\n",
       "      <td>0.005373</td>\n",
       "      <td>0.028723</td>\n",
       "      <td>0.020981</td>\n",
       "      <td>0.000417</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.018627</td>\n",
       "      <td>0.036220</td>\n",
       "      <td>0.015588</td>\n",
       "      <td>0.021449</td>\n",
       "      <td>0.009364</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000306</td>\n",
       "      <td>0.019442</td>\n",
       "      <td>0.010386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001782</th>\n",
       "      <td>0.124927</td>\n",
       "      <td>0.195199</td>\n",
       "      <td>0.076336</td>\n",
       "      <td>0.044751</td>\n",
       "      <td>0.054878</td>\n",
       "      <td>0.032744</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.032598</td>\n",
       "      <td>0.037412</td>\n",
       "      <td>0.005402</td>\n",
       "      <td>0.007822</td>\n",
       "      <td>0.012851</td>\n",
       "      <td>0.038855</td>\n",
       "      <td>0.019556</td>\n",
       "      <td>0.014065</td>\n",
       "      <td>0.016408</td>\n",
       "      <td>0.095361</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>0.011295</td>\n",
       "      <td>0.008648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001783</th>\n",
       "      <td>0.124150</td>\n",
       "      <td>0.217632</td>\n",
       "      <td>0.084494</td>\n",
       "      <td>0.027841</td>\n",
       "      <td>0.041561</td>\n",
       "      <td>0.033910</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.024572</td>\n",
       "      <td>0.047010</td>\n",
       "      <td>0.000873</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>0.010859</td>\n",
       "      <td>0.048427</td>\n",
       "      <td>0.011122</td>\n",
       "      <td>0.015233</td>\n",
       "      <td>0.032173</td>\n",
       "      <td>0.046831</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.003185</td>\n",
       "      <td>0.012215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001784</th>\n",
       "      <td>0.172932</td>\n",
       "      <td>0.119267</td>\n",
       "      <td>0.074408</td>\n",
       "      <td>0.044969</td>\n",
       "      <td>0.007456</td>\n",
       "      <td>0.005313</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.021598</td>\n",
       "      <td>0.023347</td>\n",
       "      <td>0.025748</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.012794</td>\n",
       "      <td>0.044210</td>\n",
       "      <td>0.020503</td>\n",
       "      <td>0.009367</td>\n",
       "      <td>0.002990</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.004381</td>\n",
       "      <td>0.019837</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Lachnospiraceae  Faecalibacterium_prausnitzii  \\\n",
       "Sample Name                                                    \n",
       "G212_SS001685         0.109473                      0.244980   \n",
       "G212_SS001781         0.124860                      0.061379   \n",
       "G212_SS001782         0.124927                      0.195199   \n",
       "G212_SS001783         0.124150                      0.217632   \n",
       "G212_SS001784         0.172932                      0.119267   \n",
       "\n",
       "               Dorea_longicatena  Fusicatenibacter_saccharivorans  \\\n",
       "Sample Name                                                         \n",
       "G212_SS001685           0.044713                         0.023732   \n",
       "G212_SS001781           0.094882                         0.086011   \n",
       "G212_SS001782           0.076336                         0.044751   \n",
       "G212_SS001783           0.084494                         0.027841   \n",
       "G212_SS001784           0.074408                         0.044969   \n",
       "\n",
       "               Anaerostipes_hadrus  Bifidobacterium_adolescentis  \\\n",
       "Sample Name                                                        \n",
       "G212_SS001685             0.025135                      0.000009   \n",
       "G212_SS001781             0.033223                      0.113808   \n",
       "G212_SS001782             0.054878                      0.032744   \n",
       "G212_SS001783             0.041561                      0.033910   \n",
       "G212_SS001784             0.007456                      0.005313   \n",
       "\n",
       "               Prevotella_copri  Blautia_obeum  \\\n",
       "Sample Name                                      \n",
       "G212_SS001685          0.050613       0.013379   \n",
       "G212_SS001781          0.005373       0.028723   \n",
       "G212_SS001782          0.000003       0.032598   \n",
       "G212_SS001783          0.000004       0.024572   \n",
       "G212_SS001784          0.000003       0.021598   \n",
       "\n",
       "               _leftsquarebracket_Eubacterium_rightsquarebracket__hallii  \\\n",
       "Sample Name                                                                \n",
       "G212_SS001685                                           0.012065           \n",
       "G212_SS001781                                           0.020981           \n",
       "G212_SS001782                                           0.037412           \n",
       "G212_SS001783                                           0.047010           \n",
       "G212_SS001784                                           0.023347           \n",
       "\n",
       "               Roseburia_faecis  Akkermansia_muciniphila  \\\n",
       "Sample Name                                                \n",
       "G212_SS001685          0.109579                 0.000009   \n",
       "G212_SS001781          0.000417                 0.000002   \n",
       "G212_SS001782          0.005402                 0.007822   \n",
       "G212_SS001783          0.000873                 0.000865   \n",
       "G212_SS001784          0.025748                 0.000025   \n",
       "\n",
       "               Bifidobacterium_longum  Coprococcus_comes  Blautia_wexlerae  \\\n",
       "Sample Name                                                                  \n",
       "G212_SS001685                0.000053           0.007787          0.008149   \n",
       "G212_SS001781                0.018627           0.036220          0.015588   \n",
       "G212_SS001782                0.012851           0.038855          0.019556   \n",
       "G212_SS001783                0.010859           0.048427          0.011122   \n",
       "G212_SS001784                0.012794           0.044210          0.020503   \n",
       "\n",
       "               Collinsella_aerofaciens  Gemmiger_formicilis  \\\n",
       "Sample Name                                                   \n",
       "G212_SS001685                 0.000009             0.036900   \n",
       "G212_SS001781                 0.021449             0.009364   \n",
       "G212_SS001782                 0.014065             0.016408   \n",
       "G212_SS001783                 0.015233             0.032173   \n",
       "G212_SS001784                 0.009367             0.002990   \n",
       "\n",
       "               Ruminococcus_bicirculans  Bifidobacterium_bifidum  \\\n",
       "Sample Name                                                        \n",
       "G212_SS001685                  0.008722                 0.000009   \n",
       "G212_SS001781                  0.000025                 0.000306   \n",
       "G212_SS001782                  0.095361                 0.000089   \n",
       "G212_SS001783                  0.046831                 0.000061   \n",
       "G212_SS001784                  0.000008                 0.000055   \n",
       "\n",
       "               Roseburia_intestinalis  \\\n",
       "Sample Name                             \n",
       "G212_SS001685                0.008828   \n",
       "G212_SS001781                0.019442   \n",
       "G212_SS001782                0.011295   \n",
       "G212_SS001783                0.003185   \n",
       "G212_SS001784                0.004381   \n",
       "\n",
       "               _leftsquarebracket_Ruminococcus_rightsquarebracket__torques  \n",
       "Sample Name                                                                 \n",
       "G212_SS001685                                           0.005027            \n",
       "G212_SS001781                                           0.010386            \n",
       "G212_SS001782                                           0.008648            \n",
       "G212_SS001783                                           0.012215            \n",
       "G212_SS001784                                           0.019837            "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topX = 20\n",
    "df_16S_T = df_16S.loc[df_meta.index].T\n",
    "df_16S_T['mean'] = df_16S_T.mean(axis=1)\n",
    "df_16S_T = df_16S_T.sort_values(by=['mean'],axis=0,ascending=False)\n",
    "df_16S_T = df_16S_T.drop('mean', axis=1)\n",
    "df_16S_topX = df_16S_T.iloc[0:topX].T\n",
    "df_16S_topX.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Lachnospiraceae',\n",
       " 'Faecalibacterium_prausnitzii',\n",
       " 'Dorea_longicatena',\n",
       " 'Fusicatenibacter_saccharivorans',\n",
       " 'Anaerostipes_hadrus',\n",
       " 'Bifidobacterium_adolescentis',\n",
       " 'Prevotella_copri',\n",
       " 'Blautia_obeum',\n",
       " '_leftsquarebracket_Eubacterium_rightsquarebracket__hallii',\n",
       " 'Roseburia_faecis',\n",
       " 'Akkermansia_muciniphila',\n",
       " 'Bifidobacterium_longum',\n",
       " 'Coprococcus_comes',\n",
       " 'Blautia_wexlerae',\n",
       " 'Collinsella_aerofaciens',\n",
       " 'Gemmiger_formicilis',\n",
       " 'Ruminococcus_bicirculans',\n",
       " 'Bifidobacterium_bifidum',\n",
       " 'Roseburia_intestinalis',\n",
       " '_leftsquarebracket_Ruminococcus_rightsquarebracket__torques']"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_16S_topX.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize max value of 16S data to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Faecalibacterium_prausnitzii</th>\n",
       "      <th>Dorea_longicatena</th>\n",
       "      <th>Fusicatenibacter_saccharivorans</th>\n",
       "      <th>Anaerostipes_hadrus</th>\n",
       "      <th>Bifidobacterium_adolescentis</th>\n",
       "      <th>Prevotella_copri</th>\n",
       "      <th>...</th>\n",
       "      <th>Akkermansia_muciniphila</th>\n",
       "      <th>Bifidobacterium_longum</th>\n",
       "      <th>Coprococcus_comes</th>\n",
       "      <th>Blautia_wexlerae</th>\n",
       "      <th>Collinsella_aerofaciens</th>\n",
       "      <th>Gemmiger_formicilis</th>\n",
       "      <th>Ruminococcus_bicirculans</th>\n",
       "      <th>Bifidobacterium_bifidum</th>\n",
       "      <th>Roseburia_intestinalis</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__torques</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G212_SS001685</th>\n",
       "      <td>101-030</td>\n",
       "      <td>50.135417</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.160302</td>\n",
       "      <td>0.358727</td>\n",
       "      <td>0.065474</td>\n",
       "      <td>0.034752</td>\n",
       "      <td>0.036805</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.074114</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.011403</td>\n",
       "      <td>0.011933</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.054032</td>\n",
       "      <td>0.012772</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.012927</td>\n",
       "      <td>0.007361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001781</th>\n",
       "      <td>101-007</td>\n",
       "      <td>50.052778</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.182834</td>\n",
       "      <td>0.089877</td>\n",
       "      <td>0.138937</td>\n",
       "      <td>0.125947</td>\n",
       "      <td>0.048649</td>\n",
       "      <td>0.166651</td>\n",
       "      <td>0.007868</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.027276</td>\n",
       "      <td>0.053037</td>\n",
       "      <td>0.022825</td>\n",
       "      <td>0.031407</td>\n",
       "      <td>0.013712</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.000447</td>\n",
       "      <td>0.028469</td>\n",
       "      <td>0.015208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001782</th>\n",
       "      <td>101-028</td>\n",
       "      <td>41.486111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.182932</td>\n",
       "      <td>0.285832</td>\n",
       "      <td>0.111779</td>\n",
       "      <td>0.065530</td>\n",
       "      <td>0.080359</td>\n",
       "      <td>0.047947</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011454</td>\n",
       "      <td>0.018818</td>\n",
       "      <td>0.056896</td>\n",
       "      <td>0.028636</td>\n",
       "      <td>0.020595</td>\n",
       "      <td>0.024026</td>\n",
       "      <td>0.139638</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.016540</td>\n",
       "      <td>0.012663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001783</th>\n",
       "      <td>101-028</td>\n",
       "      <td>44.583333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.181794</td>\n",
       "      <td>0.318681</td>\n",
       "      <td>0.123725</td>\n",
       "      <td>0.040768</td>\n",
       "      <td>0.060858</td>\n",
       "      <td>0.049655</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001266</td>\n",
       "      <td>0.015901</td>\n",
       "      <td>0.070912</td>\n",
       "      <td>0.016286</td>\n",
       "      <td>0.022306</td>\n",
       "      <td>0.047111</td>\n",
       "      <td>0.068575</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>0.004664</td>\n",
       "      <td>0.017887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001784</th>\n",
       "      <td>101-039</td>\n",
       "      <td>22.177083</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.253227</td>\n",
       "      <td>0.174644</td>\n",
       "      <td>0.108956</td>\n",
       "      <td>0.065849</td>\n",
       "      <td>0.010918</td>\n",
       "      <td>0.007781</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.018735</td>\n",
       "      <td>0.064737</td>\n",
       "      <td>0.030023</td>\n",
       "      <td>0.013717</td>\n",
       "      <td>0.004378</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.006415</td>\n",
       "      <td>0.029048</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Subject        Day  Diet  Lachnospiraceae  \\\n",
       "Sample Name                                                \n",
       "G212_SS001685  101-030  50.135417   0.0         0.160302   \n",
       "G212_SS001781  101-007  50.052778   0.0         0.182834   \n",
       "G212_SS001782  101-028  41.486111   0.0         0.182932   \n",
       "G212_SS001783  101-028  44.583333   0.0         0.181794   \n",
       "G212_SS001784  101-039  22.177083   2.5         0.253227   \n",
       "\n",
       "               Faecalibacterium_prausnitzii  Dorea_longicatena  \\\n",
       "Sample Name                                                      \n",
       "G212_SS001685                      0.358727           0.065474   \n",
       "G212_SS001781                      0.089877           0.138937   \n",
       "G212_SS001782                      0.285832           0.111779   \n",
       "G212_SS001783                      0.318681           0.123725   \n",
       "G212_SS001784                      0.174644           0.108956   \n",
       "\n",
       "               Fusicatenibacter_saccharivorans  Anaerostipes_hadrus  \\\n",
       "Sample Name                                                           \n",
       "G212_SS001685                         0.034752             0.036805   \n",
       "G212_SS001781                         0.125947             0.048649   \n",
       "G212_SS001782                         0.065530             0.080359   \n",
       "G212_SS001783                         0.040768             0.060858   \n",
       "G212_SS001784                         0.065849             0.010918   \n",
       "\n",
       "               Bifidobacterium_adolescentis  Prevotella_copri  ...  \\\n",
       "Sample Name                                                    ...   \n",
       "G212_SS001685                      0.000013          0.074114  ...   \n",
       "G212_SS001781                      0.166651          0.007868  ...   \n",
       "G212_SS001782                      0.047947          0.000004  ...   \n",
       "G212_SS001783                      0.049655          0.000006  ...   \n",
       "G212_SS001784                      0.007781          0.000004  ...   \n",
       "\n",
       "               Akkermansia_muciniphila  Bifidobacterium_longum  \\\n",
       "Sample Name                                                      \n",
       "G212_SS001685                 0.000013                0.000077   \n",
       "G212_SS001781                 0.000003                0.027276   \n",
       "G212_SS001782                 0.011454                0.018818   \n",
       "G212_SS001783                 0.001266                0.015901   \n",
       "G212_SS001784                 0.000036                0.018735   \n",
       "\n",
       "               Coprococcus_comes  Blautia_wexlerae  Collinsella_aerofaciens  \\\n",
       "Sample Name                                                                   \n",
       "G212_SS001685           0.011403          0.011933                 0.000013   \n",
       "G212_SS001781           0.053037          0.022825                 0.031407   \n",
       "G212_SS001782           0.056896          0.028636                 0.020595   \n",
       "G212_SS001783           0.070912          0.016286                 0.022306   \n",
       "G212_SS001784           0.064737          0.030023                 0.013717   \n",
       "\n",
       "               Gemmiger_formicilis  Ruminococcus_bicirculans  \\\n",
       "Sample Name                                                    \n",
       "G212_SS001685             0.054032                  0.012772   \n",
       "G212_SS001781             0.013712                  0.000037   \n",
       "G212_SS001782             0.024026                  0.139638   \n",
       "G212_SS001783             0.047111                  0.068575   \n",
       "G212_SS001784             0.004378                  0.000012   \n",
       "\n",
       "               Bifidobacterium_bifidum  Roseburia_intestinalis  \\\n",
       "Sample Name                                                      \n",
       "G212_SS001685                 0.000013                0.012927   \n",
       "G212_SS001781                 0.000447                0.028469   \n",
       "G212_SS001782                 0.000131                0.016540   \n",
       "G212_SS001783                 0.000089                0.004664   \n",
       "G212_SS001784                 0.000081                0.006415   \n",
       "\n",
       "               _leftsquarebracket_Ruminococcus_rightsquarebracket__torques  \n",
       "Sample Name                                                                 \n",
       "G212_SS001685                                           0.007361            \n",
       "G212_SS001781                                           0.015208            \n",
       "G212_SS001782                                           0.012663            \n",
       "G212_SS001783                                           0.017887            \n",
       "G212_SS001784                                           0.029048            \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterial_taxa = list(df_16S_topX.columns)\n",
    "df_16S_topX_w_meta = pd.merge(df_meta, df_16S_topX/df_16S_topX.max().max(), left_index=True, right_index=True, how='inner')\n",
    "df_16S_topX_w_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate log-derivatives (dlog(B)/dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Faecalibacterium_prausnitzii</th>\n",
       "      <th>Dorea_longicatena</th>\n",
       "      <th>Fusicatenibacter_saccharivorans</th>\n",
       "      <th>Anaerostipes_hadrus</th>\n",
       "      <th>Bifidobacterium_adolescentis</th>\n",
       "      <th>Prevotella_copri</th>\n",
       "      <th>...</th>\n",
       "      <th>Akkermansia_muciniphila</th>\n",
       "      <th>Bifidobacterium_longum</th>\n",
       "      <th>Coprococcus_comes</th>\n",
       "      <th>Blautia_wexlerae</th>\n",
       "      <th>Collinsella_aerofaciens</th>\n",
       "      <th>Gemmiger_formicilis</th>\n",
       "      <th>Ruminococcus_bicirculans</th>\n",
       "      <th>Bifidobacterium_bifidum</th>\n",
       "      <th>Roseburia_intestinalis</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__torques</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G212_SS001685</th>\n",
       "      <td>101-030</td>\n",
       "      <td>50.135417</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.079955</td>\n",
       "      <td>-0.305886</td>\n",
       "      <td>0.102930</td>\n",
       "      <td>-0.013917</td>\n",
       "      <td>0.105432</td>\n",
       "      <td>-0.000003</td>\n",
       "      <td>3.156090e-03</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>-0.000045</td>\n",
       "      <td>0.004126</td>\n",
       "      <td>0.010843</td>\n",
       "      <td>-0.000032</td>\n",
       "      <td>-0.078115</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>-0.000018</td>\n",
       "      <td>0.027831</td>\n",
       "      <td>0.015173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001781</th>\n",
       "      <td>101-007</td>\n",
       "      <td>50.052778</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.350997</td>\n",
       "      <td>0.228537</td>\n",
       "      <td>-0.160566</td>\n",
       "      <td>-0.106509</td>\n",
       "      <td>-0.013696</td>\n",
       "      <td>-0.119875</td>\n",
       "      <td>1.004450e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>-0.053520</td>\n",
       "      <td>-0.060096</td>\n",
       "      <td>-0.019855</td>\n",
       "      <td>-0.048225</td>\n",
       "      <td>-0.001002</td>\n",
       "      <td>-0.000118</td>\n",
       "      <td>-0.001719</td>\n",
       "      <td>0.149186</td>\n",
       "      <td>-0.015777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001782</th>\n",
       "      <td>101-028</td>\n",
       "      <td>41.486111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.033039</td>\n",
       "      <td>0.213033</td>\n",
       "      <td>0.028208</td>\n",
       "      <td>0.046750</td>\n",
       "      <td>-0.001887</td>\n",
       "      <td>0.008359</td>\n",
       "      <td>1.272122e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.009206</td>\n",
       "      <td>0.007687</td>\n",
       "      <td>0.011981</td>\n",
       "      <td>0.013598</td>\n",
       "      <td>0.010594</td>\n",
       "      <td>0.019558</td>\n",
       "      <td>-0.049944</td>\n",
       "      <td>-0.000054</td>\n",
       "      <td>0.010307</td>\n",
       "      <td>-0.007841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001783</th>\n",
       "      <td>101-028</td>\n",
       "      <td>44.583333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.006915</td>\n",
       "      <td>0.040284</td>\n",
       "      <td>0.028467</td>\n",
       "      <td>0.013208</td>\n",
       "      <td>0.012539</td>\n",
       "      <td>-0.012373</td>\n",
       "      <td>8.850795e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.005774</td>\n",
       "      <td>0.000416</td>\n",
       "      <td>0.010895</td>\n",
       "      <td>0.007495</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.001751</td>\n",
       "      <td>-0.068225</td>\n",
       "      <td>-0.000055</td>\n",
       "      <td>0.026606</td>\n",
       "      <td>0.006785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G212_SS001784</th>\n",
       "      <td>101-039</td>\n",
       "      <td>22.177083</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.322493</td>\n",
       "      <td>-0.026923</td>\n",
       "      <td>0.136809</td>\n",
       "      <td>-0.006970</td>\n",
       "      <td>-0.040000</td>\n",
       "      <td>0.003048</td>\n",
       "      <td>-5.017792e-06</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000190</td>\n",
       "      <td>0.047185</td>\n",
       "      <td>0.071169</td>\n",
       "      <td>0.001243</td>\n",
       "      <td>0.006736</td>\n",
       "      <td>-0.003201</td>\n",
       "      <td>-0.000011</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>-0.003610</td>\n",
       "      <td>-0.104799</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Subject        Day  Diet  Lachnospiraceae  \\\n",
       "Sample Name                                                \n",
       "G212_SS001685  101-030  50.135417   0.0         0.079955   \n",
       "G212_SS001781  101-007  50.052778   0.0        -0.350997   \n",
       "G212_SS001782  101-028  41.486111   0.0        -0.033039   \n",
       "G212_SS001783  101-028  44.583333   0.0        -0.006915   \n",
       "G212_SS001784  101-039  22.177083   2.5         0.322493   \n",
       "\n",
       "               Faecalibacterium_prausnitzii  Dorea_longicatena  \\\n",
       "Sample Name                                                      \n",
       "G212_SS001685                     -0.305886           0.102930   \n",
       "G212_SS001781                      0.228537          -0.160566   \n",
       "G212_SS001782                      0.213033           0.028208   \n",
       "G212_SS001783                      0.040284           0.028467   \n",
       "G212_SS001784                     -0.026923           0.136809   \n",
       "\n",
       "               Fusicatenibacter_saccharivorans  Anaerostipes_hadrus  \\\n",
       "Sample Name                                                           \n",
       "G212_SS001685                        -0.013917             0.105432   \n",
       "G212_SS001781                        -0.106509            -0.013696   \n",
       "G212_SS001782                         0.046750            -0.001887   \n",
       "G212_SS001783                         0.013208             0.012539   \n",
       "G212_SS001784                        -0.006970            -0.040000   \n",
       "\n",
       "               Bifidobacterium_adolescentis  Prevotella_copri  ...  \\\n",
       "Sample Name                                                    ...   \n",
       "G212_SS001685                     -0.000003      3.156090e-03  ...   \n",
       "G212_SS001781                     -0.119875      1.004450e+00  ...   \n",
       "G212_SS001782                      0.008359      1.272122e-08  ...   \n",
       "G212_SS001783                     -0.012373      8.850795e-07  ...   \n",
       "G212_SS001784                      0.003048     -5.017792e-06  ...   \n",
       "\n",
       "               Akkermansia_muciniphila  Bifidobacterium_longum  \\\n",
       "Sample Name                                                      \n",
       "G212_SS001685                 0.000007               -0.000045   \n",
       "G212_SS001781                 0.000008               -0.053520   \n",
       "G212_SS001782                -0.009206                0.007687   \n",
       "G212_SS001783                -0.005774                0.000416   \n",
       "G212_SS001784                -0.000190                0.047185   \n",
       "\n",
       "               Coprococcus_comes  Blautia_wexlerae  Collinsella_aerofaciens  \\\n",
       "Sample Name                                                                   \n",
       "G212_SS001685           0.004126          0.010843                -0.000032   \n",
       "G212_SS001781          -0.060096         -0.019855                -0.048225   \n",
       "G212_SS001782           0.011981          0.013598                 0.010594   \n",
       "G212_SS001783           0.010895          0.007495                 0.000309   \n",
       "G212_SS001784           0.071169          0.001243                 0.006736   \n",
       "\n",
       "               Gemmiger_formicilis  Ruminococcus_bicirculans  \\\n",
       "Sample Name                                                    \n",
       "G212_SS001685            -0.078115                  0.000114   \n",
       "G212_SS001781            -0.001002                 -0.000118   \n",
       "G212_SS001782             0.019558                 -0.049944   \n",
       "G212_SS001783             0.001751                 -0.068225   \n",
       "G212_SS001784            -0.003201                 -0.000011   \n",
       "\n",
       "               Bifidobacterium_bifidum  Roseburia_intestinalis  \\\n",
       "Sample Name                                                      \n",
       "G212_SS001685                -0.000018                0.027831   \n",
       "G212_SS001781                -0.001719                0.149186   \n",
       "G212_SS001782                -0.000054                0.010307   \n",
       "G212_SS001783                -0.000055                0.026606   \n",
       "G212_SS001784                 0.000102               -0.003610   \n",
       "\n",
       "               _leftsquarebracket_Ruminococcus_rightsquarebracket__torques  \n",
       "Sample Name                                                                 \n",
       "G212_SS001685                                           0.015173            \n",
       "G212_SS001781                                          -0.015777            \n",
       "G212_SS001782                                          -0.007841            \n",
       "G212_SS001783                                           0.006785            \n",
       "G212_SS001784                                          -0.104799            \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_deriv = deepcopy(df_16S_topX_w_meta)\n",
    "for curr_sub in set(df_deriv.Subject):\n",
    "    curr_df = df_deriv[df_deriv.Subject==curr_sub].sort_values(by='Day')\n",
    "    for taxon in bacterial_taxa:\n",
    "        xdata = np.array(curr_df['Day'])\n",
    "        ydata = np.array(curr_df[taxon])\n",
    "        cs = CubicSpline(xdata, ydata)\n",
    "        csd1 = cs.derivative(nu=1)\n",
    "        ydata_d1 = csd1(xdata)\n",
    "        df_deriv.loc[df_deriv.Subject==curr_sub, taxon] = ydata_d1\n",
    "df_deriv.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct regression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0.89394064, -3.99639251, -0.38828353, ..., -0.01298858,\n",
       "        0.0542411 ,  1.18126258])"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ymat should be samples by bacteria\n",
    "Ymat = df_deriv[bacterial_taxa].values\n",
    "Ymat = Ymat.flatten(order='F')\n",
    "Ymat = StandardScaler().fit_transform(Ymat.reshape(-1,1)).reshape(1,-1)[0] # standardize\n",
    "Ymat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18920"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Ymat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.00000000e+00, 0.00000000e+00, 1.60302274e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 0.00000000e+00, 1.82833722e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 0.00000000e+00, 1.82931605e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       ...,\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        2.36232233e-02, 1.59073939e-03, 1.96240747e-03],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        4.33919309e-06, 4.68198934e-03, 1.49485202e-02],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        8.91873007e-06, 5.12113480e-02, 5.09259487e-02]])"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Xmat = np.zeros(shape=(topX*len(df_deriv.index), (topX+2)*topX))\n",
    "for k in np.arange(topX):\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)] = 1\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+1] = df_deriv.Diet.values\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+2:(k+1)*(topX+2)] = df_16S_topX_w_meta[bacterial_taxa].values\n",
    "Xmat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write stan program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "json_str = '{\\n\"N\" : %d,\\n'%(len(Ymat))\n",
    "\n",
    "# y variable\n",
    "json_str += '\\\"dlogX\\\" : [%s],\\n'%(','.join(list(Ymat.astype(str))))\n",
    "\n",
    "# x variable\n",
    "for k1,c1 in enumerate(bacterial_taxa):\n",
    "    # growth rate\n",
    "    json_str += '\\\"growth_rate_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)].astype(str))))\n",
    "    # diet response\n",
    "    json_str += '\\\"rs_response_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)+1].astype(str))))\n",
    "    # bacterial interactions\n",
    "    for k2,c2 in enumerate(bacterial_taxa):\n",
    "        v = list(Xmat[:,k1*(topX+2)+2+k2].astype(str))\n",
    "        json_str += '\\\"pairwise_interaction_%s_%s\\\" : [%s]'%(c1,c2,','.join(v))\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            json_str += '\\n}'\n",
    "        else:\n",
    "            json_str += ',\\n'\n",
    "#print(json_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.data.json\", \"w\")\n",
    "text_file.write(\"%s\" % json_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# data block\n",
    "model_str = 'data {\\n'\n",
    "model_str += '\\tint<lower=0> N;\\n'\n",
    "model_str += '\\tvector[N] dlogX;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\tvector[N] growth_rate_%s;\\n'%(c1)\n",
    "    model_str += '\\tvector[N] rs_response_%s;\\n'%(c1)\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tvector[N] pairwise_interaction_%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'\n",
    "\n",
    "# parameter block\n",
    "model_str += 'parameters {\\n\\treal<lower=0,upper=1> sigma;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\treal alpha__%s;\\n'%(c1) # growth rate\n",
    "    model_str += '\\treal epsilon__%s;\\n'%(c1) # rs response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\treal beta__%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'       \n",
    "        \n",
    "# model block\n",
    "model_str += 'model {\\n\\tsigma ~ uniform(0,1);\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\talpha__%s ~ normal(0,1);\\n'%(c1) # growth rate\n",
    "    model_str += '\\tepsilon__%s ~ normal(0,1);\\n'%(c1) # rs response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tbeta__%s_%s ~ normal(0,1);\\n'%(c1,c2)\n",
    "model_str += '\\tdlogX ~ normal('\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += 'alpha__%s*growth_rate_%s+'%(c1,c1) # growth rate\n",
    "    model_str += 'epsilon__%s*rs_response_%s+'%(c1,c1) # rs response\n",
    "    for c2 in bacterial_taxa:\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s'%(c1,c2,c1,c2)\n",
    "        else:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s+'%(c1,c2,c1,c2)\n",
    "model_str += ', sigma);\\n}'\n",
    "#print(model_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.stan\", \"w\")\n",
    "text_file.write(\"%s\" % model_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot stan output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hpd_grid(sample, alpha=0.05, roundto=2):\n",
    "    \"\"\"Calculate highest posterior density (HPD) of array for given alpha. \n",
    "    The HPD is the minimum width Bayesian credible interval (BCI). \n",
    "    The function works for multimodal distributions, returning more than one mode\n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    sample : Numpy array or python list\n",
    "        An array containing MCMC samples\n",
    "    alpha : float\n",
    "        Desired probability of type I error (defaults to 0.05)\n",
    "    roundto: integer\n",
    "        Number of digits after the decimal point for the results\n",
    "    Returns\n",
    "    ----------\n",
    "    hpd: array with the lower \n",
    "          \n",
    "    \"\"\"\n",
    "    sample = np.asarray(sample)\n",
    "    sample = sample[~np.isnan(sample)]\n",
    "    # get upper and lower bounds\n",
    "    l = np.min(sample)\n",
    "    u = np.max(sample)\n",
    "    density = kde.gaussian_kde(sample)\n",
    "    x = np.linspace(l, u, 2000)\n",
    "    y = density.evaluate(x)\n",
    "    #y = density.evaluate(x, l, u) waitting for PR to be accepted\n",
    "    xy_zipped = zip(x, y/np.sum(y))\n",
    "    xy = sorted(xy_zipped, key=lambda x: x[1], reverse=True)\n",
    "    xy_cum_sum = 0\n",
    "    hdv = []\n",
    "    for val in xy:\n",
    "        xy_cum_sum += val[1]\n",
    "        hdv.append(val[0])\n",
    "        if xy_cum_sum >= (1-alpha):\n",
    "            break\n",
    "    hdv.sort()\n",
    "    diff = (u-l)/20  # differences of 5%\n",
    "    hpd = []\n",
    "    hpd.append(round(min(hdv), roundto))\n",
    "    for i in range(1, len(hdv)):\n",
    "        if hdv[i]-hdv[i-1] >= diff:\n",
    "            hpd.append(round(hdv[i-1], roundto))\n",
    "            hpd.append(round(hdv[i], roundto))\n",
    "    hpd.append(round(max(hdv), roundto))\n",
    "    ite = iter(hpd)\n",
    "    hpd = list(zip(ite, ite))\n",
    "    modes = []\n",
    "    for value in hpd:\n",
    "        x_hpd = x[(x > value[0]) & (x < value[1])]\n",
    "        y_hpd = y[(x > value[0]) & (x < value[1])]\n",
    "        modes.append(round(x_hpd[np.argmax(y_hpd)], roundto))\n",
    "    return hpd, x, y, modes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = az.from_cmdstan([\"output_%d.csv\"%(i) for i in np.arange(1,4)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>-0.42</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.159749</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "      <td>-0.60</td>\n",
       "      <td>0.32</td>\n",
       "      <td>-0.140</td>\n",
       "      <td>-0.605190</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>-0.66</td>\n",
       "      <td>0.26</td>\n",
       "      <td>-0.200</td>\n",
       "      <td>-0.856932</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>gut_dash_metagenome</td>\n",
       "      <td>-0.14</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.320</td>\n",
       "      <td>1.414249</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das...</td>\n",
       "      <td>-0.34</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.115</td>\n",
       "      <td>0.486693</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Type                                               Taxa  Left  \\\n",
       "0  basal_growth_rate                                        Bacteroides -0.42   \n",
       "1  basal_growth_rate                                   Faecalibacterium -0.60   \n",
       "2  basal_growth_rate                                    Lachnospiraceae -0.66   \n",
       "3  basal_growth_rate                                gut_dash_metagenome -0.14   \n",
       "4  basal_growth_rate  Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das... -0.34   \n",
       "\n",
       "   Right  Middle       SNR  Significant  \n",
       "0   0.50   0.040  0.159749        False  \n",
       "1   0.32  -0.140 -0.605190        False  \n",
       "2   0.26  -0.200 -0.856932        False  \n",
       "3   0.78   0.320  1.414249        False  \n",
       "4   0.57   0.115  0.486693        False  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines = []\n",
    "    \n",
    "# basal growth rate\n",
    "for c in bacterial_taxa:\n",
    "    var = 'alpha__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['basal_growth_rate', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# inulin response\n",
    "for c in bacterial_taxa:\n",
    "    var = 'epsilon__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['rs_response', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "\n",
    "# pairwise interactions\n",
    "for c1 in bacterial_taxa:\n",
    "    for c2 in bacterial_taxa:\n",
    "        var = 'beta__%s_%s'%(c1,c2)\n",
    "        data = []\n",
    "        for i in np.arange(0,3):\n",
    "            data.extend(list(fit.posterior[var][i].values))\n",
    "        hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "        assert len(hpd_mu) == 1\n",
    "        (x0, x1) = hpd_mu[0]\n",
    "        lines.append(['pairwise_interaction', (c1,c2), x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "df_stan_output_summary = pd.DataFrame(lines, columns = ['Type','Taxa','Left','Right','Middle','SNR','Significant'])\n",
    "df_stan_output_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stan_output_summary.to_excel('bayesian_regression_summary_top20.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.405</td>\n",
       "      <td>2.684754</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das...</td>\n",
       "      <td>-0.64</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>-0.345</td>\n",
       "      <td>-2.313357</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Anaerostipes</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.400</td>\n",
       "      <td>2.642284</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Bifidobacterium</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.485</td>\n",
       "      <td>3.229673</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Bacteroides')</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1.270</td>\n",
       "      <td>3.432835</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Faecalibacterium')</td>\n",
       "      <td>-7.65</td>\n",
       "      <td>-4.91</td>\n",
       "      <td>-6.280</td>\n",
       "      <td>-9.042536</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'gut_dash_metagenome')</td>\n",
       "      <td>0.02</td>\n",
       "      <td>3.46</td>\n",
       "      <td>1.740</td>\n",
       "      <td>1.989138</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Ruminococcus_dash_sp_dot__das...</td>\n",
       "      <td>0.51</td>\n",
       "      <td>4.05</td>\n",
       "      <td>2.280</td>\n",
       "      <td>2.536649</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Blautia')</td>\n",
       "      <td>-3.70</td>\n",
       "      <td>-0.08</td>\n",
       "      <td>-1.890</td>\n",
       "      <td>-2.059112</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Faecalibacterium', 'Faecalibacterium')</td>\n",
       "      <td>1.53</td>\n",
       "      <td>4.27</td>\n",
       "      <td>2.900</td>\n",
       "      <td>4.187724</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bifidobacterium', 'Bifidobacterium')</td>\n",
       "      <td>0.31</td>\n",
       "      <td>3.93</td>\n",
       "      <td>2.120</td>\n",
       "      <td>2.333634</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Type                                               Taxa  \\\n",
       "22            rs_response                                    Lachnospiraceae   \n",
       "24            rs_response  Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das...   \n",
       "29            rs_response                                       Anaerostipes   \n",
       "31            rs_response                                    Bifidobacterium   \n",
       "40   pairwise_interaction                     ('Bacteroides', 'Bacteroides')   \n",
       "41   pairwise_interaction                ('Bacteroides', 'Faecalibacterium')   \n",
       "43   pairwise_interaction             ('Bacteroides', 'gut_dash_metagenome')   \n",
       "44   pairwise_interaction  ('Bacteroides', 'Ruminococcus_dash_sp_dot__das...   \n",
       "46   pairwise_interaction                         ('Bacteroides', 'Blautia')   \n",
       "61   pairwise_interaction           ('Faecalibacterium', 'Faecalibacterium')   \n",
       "271  pairwise_interaction             ('Bifidobacterium', 'Bifidobacterium')   \n",
       "\n",
       "     Left  Right  Middle       SNR  Significant  \n",
       "22   0.11   0.70   0.405  2.684754         True  \n",
       "24  -0.64  -0.05  -0.345 -2.313357         True  \n",
       "29   0.10   0.70   0.400  2.642284         True  \n",
       "31   0.19   0.78   0.485  3.229673         True  \n",
       "40   0.54   2.00   1.270  3.432835         True  \n",
       "41  -7.65  -4.91  -6.280 -9.042536         True  \n",
       "43   0.02   3.46   1.740  1.989138         True  \n",
       "44   0.51   4.05   2.280  2.536649         True  \n",
       "46  -3.70  -0.08  -1.890 -2.059112         True  \n",
       "61   1.53   4.27   2.900  4.187724         True  \n",
       "271  0.31   3.93   2.120  2.333634         True  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stan_output_summary = pd.read_excel('bayesian_regression_summary_top20.xlsx', index_col=0)\n",
    "df_stan_output_summary = df_stan_output_summary[df_stan_output_summary.Significant==True]\n",
    "df_stan_output_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
